---
title: "eiad"
author: "Lea Ahrens"
date: "28 5 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{r}
load("../2021-topic-01-team-03/dataset/prism_datasets.rda")
load("../2021-topic-01-team-03/dataset/cellline_datasets.rda")
```


# Guidelines

prism: effect of the treatment (columns) on cell growth of cell lines (rows).For example, BRD-A00077618-236-07-6::0.0024414::HTS002 indicates treatment with the drug BRD-A00077618-236-07-6, with a dose of 2.4 nM, in assay HTS002. Values represent log fold changes with respect to a control DMSO experiment.

prism.treat:information about the treatment (row) and the drug used

prism.cl: cell lines' info

prism.exp: gene TPM (transcripts per million) values. Indicate over/underexpression. Rows = cell lines, columns = genes

prism.cnv: gene copy number (CN) values. In absolute terms, CN = 2, since there are two alleles per genes. In cancer, genes might be amplified CN > 2 or deleted CN < 2

prism.snv: observed mutations in a sample. The isDeleterious flag specifies if mutation has a functional effect or not. First column is genes. There is a column
(DepMap_ID) with corresponding cell lines

prism.achilles: gene KO scores, a measure of how essential/important is a particular gene for the cell survival

IDEA: use same indications as in PRISM paper to start filtering for relevant genes, cell lines? 

# Cleaning

## Selecting brain cancer specific cell lines in all data sets

```{r}
brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") #identify the brain cancer cell lines' names
names=brain_cancer_cl[,1] #saving the names as names
#select wanted cell lines from all datasets
brain_cancer=prism[names,]
brain_cancer_achilles=prism.achilles[names,]
brain_cancer_cnv=prism.cnv[names,]
brain_cancer_exp=prism.exp[names,]
```

##Selecting relevant info from prism.snv

```{r
for (i in names){
 result<-which(prism.snv$DepMap_ID == 	
i)
}
result #only delivers results for one of 34 cell lines. without saving it as result it works on all cell lines
sapply(names, function(i){
  snv<-which(prism.snv$DepMap_ID == 	
i)
  View(prism.snv[snv,])
}) #works but kinda weird
```

## NA removal from brain_cancer_achilles

```{r}
brain_cancer_achilles=na.omit(brain_cancer_achilles) #remove the entire cell lines from everywhere without KO scores
```

## NA for mean substituition in brain_cancer

```{r}
for(i in 1:ncol(brain_cancer)){
  brain_cancer[is.na(brain_cancer[,i]), i] <- mean(brain_cancer[,i], na.rm = TRUE)
}
brain_cancer=brain_cancer[ , colSums(is.na(brain_cancer)) == 0]
```

## Dividing cell lines into subgroups

```{r}
medulloblastoma=brain_cancer_cl$DepMap_ID[grep("Medulloblastoma", brain_cancer_cl$disease_subtype)]#we obtain the names of the cell lines that are classified as medulloblastoma
glioblastoma=brain_cancer_cl$DepMap_ID[grep("Glioblastoma", brain_cancer_cl$disease_subtype)]
glioma=brain_cancer_cl$DepMap_ID[grep("Glioma", brain_cancer_cl$disease_subtype)]
astrocytoma=brain_cancer_cl$DepMap_ID[grep("Astrocytoma", brain_cancer_cl$disease_subtype)]
disease_subtypes=c(medulloblastoma,glioblastoma,glioma,astrocytoma)
length(disease_subtypes)#making sure we didnt forget anything
```

# Question 1: How can we distinguish the most effective drugs?

## Deviding drugs in doses groups

```{r}
dd1=grep("::0.00061034::", colnames(brain_cancer))
dd2=grep("::0.0024414::", colnames(brain_cancer))
dd3=grep("::0.00976562::", colnames(brain_cancer))
dd4=grep("::0.0390625::", colnames(brain_cancer))
dd5=grep("::0.15625::", colnames(brain_cancer))
dd6=grep("::0.625::", colnames(brain_cancer))
dd7=grep("::10::", colnames(brain_cancer))
dd8=grep("::2.5::", colnames(brain_cancer))
#positions of the respective doses in the dataframe 
doses=c(dd1,dd2,dd3,dd4,dd5,dd6,dd7,dd8)
d1=brain_cancer[,dd1]
d2=brain_cancer[,dd2]
d3=brain_cancer[,dd3]
d4=brain_cancer[,dd4]
d5=brain_cancer[,dd5]
d6=brain_cancer[,dd6]
d7=brain_cancer[,dd7]
d8=brain_cancer[,dd8]
#these are the standard doses summed up into seperate dataframes
sum(length(dd1), length(dd2), length(dd3), length(dd4),length(dd5), length(dd6),length(dd7),length(dd8)) #ups something missing
deviation=brain_cancer[,-doses] #lets see what we missed-> these are the doses that are not exact and have some kind of a deviation
questionable_drugs=prism.treat[colnames(deviation),]#we dont know if we can use these drugs, may be their doses deviate too far from the standard
dim(questionable_drugs)
```

## Working on drugs that do not fit the standard doses

```{r}
#Stefan's suggestion: assigning the outliers to closest standard doses
#Calculation of all differences between questionable drugs and standard doses
std_doses <- c(0.00061034,0.0024414,0.00976562,0.0390625,0.15625,0.625,10,2.5) #vector with standard dosages
#Picking dosages from dosages that have to be assigned to standard dosages
doses_to_assign=as.numeric(questionable_drugs$dose)
#Creating data frame for the differences between standard dosages and questionable drugs
extra_drugs <-data.frame(matrix(nrow = 8,ncol = length(doses_to_assign)),row.names = std_doses)
colnames(extra_drugs)=rownames(questionable_drugs)
for (i in 1:length(doses_to_assign)){
    extra_drugs[,i]=abs(std_doses - doses_to_assign[i])
}
#Assigning the drugs to the closest standard dosage 
to_which_dose=apply(extra_drugs, 2, which.min) #function of which.min: gives row number of min value of column. In comparison just apply: gives min difference
d1_extra=which(to_which_dose==1)
d2_extra=which(to_which_dose==2)
d3_extra=which(to_which_dose==3)
d4_extra=which(to_which_dose==4)
d5_extra=which(to_which_dose==5)
d6_extra=which(to_which_dose==6)
d7_extra=which(to_which_dose==7)
d8_extra=which(to_which_dose==8)
#just a quick check how many doses there are in total
not_standard_doses=c(d1_extra,d2_extra,d3_extra,d4_extra,d5_extra,d6_extra,d7_extra,d8_extra)
length(not_standard_doses)#in comparison with dimension from questionable_drugs right!
#Adding new assigned dosages to small data frames of standard dosages 
add_extra_doses=function(x,y){cbind(brain_cancer[,unique(grep(paste(colnames(extra_drugs[,x]),collapse="|"), colnames(brain_cancer)))], y)}
d1=add_extra_doses(d1_extra,d1)
d2=add_extra_doses(d2_extra,d2)
d3=add_extra_doses(d3_extra,d3)
d4=add_extra_doses(d4_extra,d4)
d5=add_extra_doses(d5_extra,d5)
d6=add_extra_doses(d6_extra,d6)
d7=add_extra_doses(d7_extra,d7)
d8=add_extra_doses(d8_extra,d8)
#Resumé: new data sets for each dosage with assigned outliers to closest deviation
ncol(d2)+ncol(d1)+ncol(d3)+ncol(d4)+ncol(d5)+ncol(d6)+ncol(d7)+ncol(d8)
ncol(brain_cancer)
```


## Filter effective drugs in each dose group

```{r}
#Lets say for now that our threshold is 0,2 so we can see how many drugs meet this criteria in each dose group
effective_drugs<-function(x){
  effective_x<-x[,which(apply(x,2,mean)<=0.2)]
  return(effective_x)
}
effective_d1=effective_drugs(d1)
effective_d2=effective_drugs(d2)
effective_d3=effective_drugs(d3)
effective_d4=effective_drugs(d4)
effective_d5=effective_drugs(d5)
effective_d6=effective_drugs(d6)
effective_d7=effective_drugs(d7)
effective_d8=effective_drugs(d8)
all_effective_drugs=c(colnames(effective_d1),colnames(effective_d2),colnames(effective_d3),colnames(effective_d4),colnames(effective_d5),colnames(effective_d6),colnames(effective_d7),colnames(effective_d8))
length(all_effective_drugs)
length(which(apply(brain_cancer,2,mean)<=0.2)) #this is the amount of drugs that we keep in total. We compare these 2 length to make sure that that we do not select the same drug twice. If it was the case the first length would be bigger than the second one
```

## Apllying efective doses filtering on prism.treat

```{r}
brain_cancer_treat=prism.treat[colnames(brain_cancer),] #selecting only the drugs we previously filtered out from brain_cancer
effective_t1=brain_cancer_treat[colnames(effective_d1),]
effective_t2=brain_cancer_treat[colnames(effective_d2),]
effective_t3=brain_cancer_treat[colnames(effective_d3),]
effective_t4=brain_cancer_treat[colnames(effective_d4),]
effective_t5=brain_cancer_treat[colnames(effective_d5),]
effective_t6=brain_cancer_treat[colnames(effective_d6),]
effective_t7=brain_cancer_treat[colnames(effective_d7),]
effective_t8=brain_cancer_treat[colnames(effective_d8),]
```

##Select drugs that are effective in all doses

```{r}
new_colnames<-function(x){
  a<-gsub(pattern = "::.*", replacement = "", colnames(x))
  return(a)
}
colnames(effective_d1)<-new_colnames(effective_d1)
colnames(effective_d2)<-new_colnames(effective_d2)
colnames(effective_d3)<-new_colnames(effective_d3)
colnames(effective_d4)<-new_colnames(effective_d4)
colnames(effective_d5)<-new_colnames(effective_d5)
colnames(effective_d6)<-new_colnames(effective_d6)
colnames(effective_d7)<-new_colnames(effective_d7)
colnames(effective_d8)<-new_colnames(effective_d8)
## remove everything but the drug information from colnames and make them new colnames of the dataframes
save_new_colnames=function(y){
  colnames(y)<-new_colnames(y)
}
save_new_colnames(effective_d4)
#does not work
  
common_drugs<-function(x,y){
  different<-setdiff(colnames(x), colnames(y))
  
  different<-unique(grep(paste(different,collapse="|"), colnames(x)))
  
  colnames(x[,-different])
}
#function returns the names of the common drugs between 2 dataframes
d1_d2=common_drugs(effective_d1,effective_d2)
d2_d1=common_drugs(effective_d2,effective_d1)
#both are equal, so the function works correctly
d1_d3=common_drugs(effective_d1,effective_d3)
d1_d4=common_drugs(effective_d1,effective_d4)
d1_d5=common_drugs(effective_d1,effective_d5)
d1_d6=common_drugs(effective_d1,effective_d6)
d1_d7=common_drugs(effective_d1,effective_d7)
d1_d8=common_drugs(effective_d1,effective_d8)
#which drugs are present in all 8 doses?
present_in_all_doses<-function(a,b,c,d,e,f,g){
 bb<-setdiff(a, b)
 a<-a [! a %in% bb]
 
 cc<-setdiff(a, c)
 a<-a [! a %in% cc]
 
 dd<-setdiff(a, d)
 a<-a [! a %in% dd]
 
 ee<-setdiff(a, e)
 a<-a [! a %in% ee]
 
 ff<-setdiff(a, f)
 a<-a [! a %in% ff]
 
 gg<-setdiff(a, g)
 a<-a [! a %in% gg]
 
 return(a)
}
effective_in_all_doses=present_in_all_doses(d1_d2,d1_d3,d1_d4,d1_d5,d1_d6,d1_d7,d1_d8)
#51 drugs are effective in all doses

```

```{r}
#Proofing...
#mean(brain_cancer$`BRD-A68969091-001-12-7::0.000628134::HTS002`)
#[1] 0.1459877
#mean(brain_cancer$`BRD-A00077618-236-07-6::0.0024414::HTS002`)
#[1] 0.2307986

```

```{r}
effective_dosis <- list(d1 <- effective_d1, d2 <- effective_d2, d3 <- effective_d3, d4 <- effective_d4, d5 <- effective_d5, d6 <- effective_d6, d7 <- effective_d7, d8 <- effective_d8)

for (i in 1:length(effective_dosis)){
  effective_dosis[[i]] <- effective_dosis[[i]][,which(colnames(effective_dosis[[i]]) %in% effective_in_all_doses)]
  effective_dosis[[i]] <- apply(effective_dosis[[i]], 2, mean)
  effective_dosis[[i]] <- as.vector(effective_dosis[[i]])
}

for (i in 1:length(effective_dosis)){
  hist(effective_dosis[[i]], main = paste("Dose", i, "of effective drugs"), xlab = "effectivity")
  q <- quantile(effective_dosis[[i]], probs = c(0.05, 0.5, 0.95))
  abline(v = q, col = "red", lty = 2, lwd = 2)
}


q_matrix <- rbind(q1,q2,q3,q4,q5,q6,q7,q8)

#heatmap
heatmap(as.matrix(mean_eiad_matrix))#--> cluster

#heatmap.2
#install.packages("gplots")
#library(gplots)
#heatmap.2(as.matrix(mean_eiad_matrix))

#histogram of each doses per drug (rows of mean_eiad)
plot(mean_eiad_matrix[1,], type = "b", col = "red", ylim = c(-8,1))#0, fällt/rollt, 4
points(mean_eiad_matrix[2,], type = "b", col = "chocolate")#0,fällt pik6
points(mean_eiad_matrix[3,], type = "b", col = "lightskyblue1")#0schwankt
points(mean_eiad_matrix[4,], type = "b", col = "green")#1, fällt, 6
points(mean_eiad_matrix[5,], type = "b", col = "darkgreen")#1, rollt, 4
points(mean_eiad_matrix[6,], type = "b", col = "magenta")#0,fällt bei 4d, 6
points(mean_eiad_matrix[7,], type = "b", col = "darkmagenta")#0, fällt bei 3d, 6
points(mean_eiad_matrix[8,], type = "b", col = "yellow")#2,schwankt um ca 2
points(mean_eiad_matrix[9,], type = "b", col = "black")#nah 0, rollt pik 2
points(mean_eiad_matrix[10,], type = "b", col = "grey")#0, rollt/fällt, 4

plot(mean_eiad_matrix[11,], type = "b", col = "red", ylim = c(-8,1))#0,fällt bei 6d,5
points(mean_eiad_matrix[12,], type = "b", col = "chocolate")#0, fällt bei 5d, 4
points(mean_eiad_matrix[13,], type = "b", col = "lightskyblue1")#0,schwankt
points(mean_eiad_matrix[14,], type = "b", col = "green")#0, fällt bei 5d, 2
points(mean_eiad_matrix[15,], type = "b", col = "darkgreen")#2, zz, fällt,4 
points(mean_eiad_matrix[16,], type = "b", col = "magenta")#0,, fällt bei 4d, 6
points(mean_eiad_matrix[17,], type = "b", col = "darkmagenta")#0, fällt bei 3d, konst
points(mean_eiad_matrix[18,], type = "b", col = "yellow")#2,schwankt light zz
points(mean_eiad_matrix[19,], type = "b", col = "black")#0, fällt bei 6d, 3
points(mean_eiad_matrix[20,], type = "b", col = "grey")#0, fällt bei 2d, 4

plot(mean_eiad_matrix[21,], type = "b", col = "red", ylim = c(-8,1))#0, fb 2d, 2
points(mean_eiad_matrix[22,], type = "b", col = "chocolate")#0, fb 6d, 2
points(mean_eiad_matrix[23,], type = "b", col = "lightskyblue1")#0, zz, 3
points(mean_eiad_matrix[24,], type = "b", col = "green")#0, light zz, fb 5d, 4
points(mean_eiad_matrix[25,], type = "b", col = "darkgreen")#2, light zz, ab 4d const.
points(mean_eiad_matrix[26,], type = "b", col = "magenta")#0, fällt leicht, 1,5
points(mean_eiad_matrix[27,], type = "b", col = "darkmagenta")#1,5, fällt, 4
points(mean_eiad_matrix[28,], type = "b", col = "yellow")# 1, zz, 6
points(mean_eiad_matrix[29,], type = "b", col = "black")#1,  light zz, 2
points(mean_eiad_matrix[30,], type = "b", col = "grey")#0, fällt bei 2d, 6

plot(mean_eiad_matrix[31,], type = "b", col = "red", ylim = c(-8,1))#0,light zz, fb 6d, 4
points(mean_eiad_matrix[32,], type = "b", col = "chocolate")#0, fb 5d, 4
points(mean_eiad_matrix[33,], type = "b", col = "lightskyblue1")#0,6d, 4
points(mean_eiad_matrix[34,], type = "b", col = "green")#0, fb 4d, 6
points(mean_eiad_matrix[35,], type = "b", col = "darkgreen")#0,fb 4d, 3
points(mean_eiad_matrix[36,], type = "b", col = "magenta")#0,fb 3d,6
points(mean_eiad_matrix[37,], type = "b", col = "darkmagenta")#0, heavy zz,5
points(mean_eiad_matrix[38,], type = "b", col = "yellow")#0, fb 3d, 5
points(mean_eiad_matrix[39,], type = "b", col = "black")#0,schwankt, 0
points(mean_eiad_matrix[40,], type = "b", col = "grey")#0,6d, 4

plot(mean_eiad_matrix[41,], type = "b", col = "red", ylim = c(-8,1))#0, fb 3d, 2
points(mean_eiad_matrix[42,], type = "b", col = "chocolate")#2, fb 6d, ca 8
points(mean_eiad_matrix[43,], type = "b", col = "lightskyblue1")#0, fällt, 6
points(mean_eiad_matrix[44,], type = "b", col = "green")#0, fällt,5
points(mean_eiad_matrix[45,], type = "b", col = "darkgreen")#0, zz, schwankt,0
points(mean_eiad_matrix[46,], type = "b", col = "magenta")#0, fb 5d, 6
points(mean_eiad_matrix[47,], type = "b", col = "darkmagenta")#2, zz, 6
points(mean_eiad_matrix[48,], type = "b", col = "yellow")#2, fb 5d, 6
points(mean_eiad_matrix[49,], type = "b", col = "black")#0, fällt, 2
points(mean_eiad_matrix[50,], type = "b", col = "grey")#0, fb 5d, pik 8d 4!!!
points(mean_eiad_matrix[51,], type = "b", col = "blue")#0, fällt, 3


###
# mean of the means?
#for (i in effective_in_all_doses)
#{mean1_160 <- mean(effective_d1[,i])}


# variation of sorting
#0, fallen, ca 4
plot(mean_eiad_matrix[10,], type = "b", col = "red", ylim = c(-8,1))#0, rollt/fällt, 4
points(mean_eiad_matrix[12,], type = "b", col = "chocolate")#0, fällt bei 5d, 4
points(mean_eiad_matrix[31,], type = "b", col = "black")#0,light zz, fb 6d, 4
points(mean_eiad_matrix[32,], type = "b", col = "chocolate")#0, fb 5d, 4
points(mean_eiad_matrix[33,], type = "b", col = "lightskyblue1")#0,6d, 4
points(mean_eiad_matrix[40,], type = "b", col = "yellow")#0,6d, 4
points(mean_eiad_matrix[17,], type = "b", col = "darkmagenta")#0, fällt bei 3d, konst

#0, fallen, 7
plot(mean_eiad_matrix[2,], type = "b", col = "red", ylim = c(-8,1))#0,fällt pik6
points(mean_eiad_matrix[6,], type = "b", col = "magenta")#0,fällt bei 4d, 6
points(mean_eiad_matrix[30,], type = "b", col = "grey")#0, fällt bei 2d, 6
points(mean_eiad_matrix[34,], type = "b", col = "green")#0, fb 4d, 6
points(mean_eiad_matrix[36,], type = "b", col = "yellow")#0,fb 3d,6
points(mean_eiad_matrix[43,], type = "b", col = "lightskyblue1")#0, fällt, 6
points(mean_eiad_matrix[46,], type = "b", col = "blue")#0, fb 5d, 6

#Schwanken um Null
plot(mean_eiad_matrix[3,], type = "b", col = "red", ylim = c(-8,1))#0schwankt
points(mean_eiad_matrix[13,], type = "b", col = "lightskyblue1")#0,schwankt
points(mean_eiad_matrix[39,], type = "b", col = "black")#0,schwankt, 0

#0, fallen, 2
plot(mean_eiad_matrix[21,], type = "b", col = "red", ylim = c(-8,1))#0, fb 2d, 2
points(mean_eiad_matrix[9,], type = "b", col = "black")#nah 0, rollt pik 2
points(mean_eiad_matrix[14,], type = "b", col = "green")#0, fällt bei 5d, 2
points(mean_eiad_matrix[22,], type = "b", col = "chocolate")#0, fb 6d, 2
points(mean_eiad_matrix[26,], type = "b", col = "magenta")#0, fällt leicht, 1,5
points(mean_eiad_matrix[41,], type = "b", col = "lightskyblue1")#0, fb 3d, 2
points(mean_eiad_matrix[49,], type = "b", col = "grey")#0, fällt, 2
points(mean_eiad_matrix[7,], type = "b", col = "darkmagenta")#0, fällt bei 3d, 2

# Peak bei dose8
plot(mean_eiad_matrix[50,], type = "b", col = "red", ylim = c(-8,1))#0, fb 5d, pik 8d 4!!!


#0, fallen, 4
plot(mean_eiad_matrix[11,], type = "b", col = "red", ylim = c(-8,1))
points(mean_eiad_matrix[38,], type = "b", col = "yellow")
points(mean_eiad_matrix[1,], type = "b", col = "lightskyblue1")
points(mean_eiad_matrix[20,], type = "b", col = "blue")

#ZickZack
plot(mean_eiad_matrix[15,], type = "b", col = "red", ylim = c(-8,1))#2, zz, fällt,4 
points(mean_eiad_matrix[18,], type = "b", col = "yellow")#2,schwankt light zz
points(mean_eiad_matrix[24,], type = "b", col = "green")#0, light zz, fb 5d, 4
points(mean_eiad_matrix[37,], type = "b", col = "darkmagenta")#0, heavy zz,5
points(mean_eiad_matrix[16,], type = "b", col = "darkgreen")#0,, fällt bei 4d, 6
points(mean_eiad_matrix[45,], type = "b", col = "blue")#
points(mean_eiad_matrix[44,], type = "b", col = "grey")

#0, fallen, ca 3
plot(mean_eiad_matrix[19,], type = "b", col = "red", ylim = c(-8,1))#0, fällt bei 6d, 3
points(mean_eiad_matrix[23,], type = "b", col = "lightskyblue1")#0, zz, 3
points(mean_eiad_matrix[35,], type = "b", col = "darkgreen")#0,fb 4d, 3
points(mean_eiad_matrix[51,], type = "b", col = "blue")#0, fällt, 3


#1
plot(mean_eiad_matrix[4,], type = "b", col = "red", ylim = c(-8,1))#1, fällt, 6
points(mean_eiad_matrix[5,], type = "b", col = "darkgreen")#1, rollt, 4
points(mean_eiad_matrix[27,], type = "b", col = "darkmagenta")#1,5, fällt, 4
points(mean_eiad_matrix[28,], type = "b", col = "yellow")# 1, zz, 6
points(mean_eiad_matrix[29,], type = "b", col = "black")#1,  light zz, 2


#2
plot(mean_eiad_matrix[8,], type = "b", col = "red", ylim = c(-8,1))#2,schwankt um ca 2
points(mean_eiad_matrix[25,], type = "b", col = "darkgreen")#2, light zz, ab 4d const.
points(mean_eiad_matrix[47,], type = "b", col = "darkmagenta")#2, zz, 6
points(mean_eiad_matrix[48,], type = "b", col = "yellow")#2, fb 5d, 6
points(mean_eiad_matrix[42,], type = "b", col = "chocolate")#2, fb 6d, ca 8

```

# Question 2: What are the targets of the effective drugs?

##What are the targets of the effective drugs?

```{r}
brain_cancer_treat=prism.treat[colnames(brain_cancer),] #selecting only the drugs we previously filtered out from brain_cancer
brain_cancer_treat=brain_cancer_treat[!duplicated(brain_cancer_treat$broad_id),]
rownames(brain_cancer_treat)<-brain_cancer_treat[,1]
effective_bc_treat=brain_cancer_treat[effective_in_all_doses,-c(1,2,3,4,10)]
#summary(effective_bc_treat$target)
#summary(effective_bc_treat$moa)
#expression of these genes would be interesting to look into our cell lines
```


Targets

```{r}
#??
#top <- which(effective_bc_treat$moa == "topoisomerase inhibitor")
# <- which(effective_bc_treat$moa == "mTOR inhibitor")
#aurk <- which(effective_bc_treat$moa == "Aurora kinase inhibitor")
#jak <- which(effective_bc_treat$moa == "JAK inhibitor")
#mek <- which(effective_bc_treat$moa == "MEK inhibitor")


```




## Drugs and subtypes relationship

```{r}
ae_d7 <-data.frame(matrix(nrow = 34,ncol = length(effective_in_all_doses)),row.names = rownames(brain_cancer))
colnames(ae_d7)=effective_in_all_doses

for (i in 1:length(effective_in_all_doses)){
  ae_d7[,i]=effective_d7[,grep(effective_in_all_doses[i], colnames(effective_d7))]
}

glioma_ae_d7=ae_d7[glioma,]

glioblastoma_ae_d7=ae_d7[glioblastoma,]

astrocytoma_ae_d7=ae_d7[astrocytoma,]

medulloblastoma_ae_d7=ae_d7[medulloblastoma,]

compare_subtypes<-data.frame(matrix(nrow = 4,ncol = length(effective_in_all_doses)),row.names = c("glioma", "glioblastoma", "astrocytoma", "medulloblastoma"))
colnames(compare_subtypes)=effective_in_all_doses

compare_subtypes[1,]<-apply(glioma_ae_d7, 2, mean)
compare_subtypes[2,]<-apply(glioblastoma_ae_d7, 2, mean)
compare_subtypes[3,]<-apply(astrocytoma_ae_d7, 2, mean)
compare_subtypes[4,]<-apply(medulloblastoma_ae_d7, 2, mean)

```

##Looking at targets of our effective drugs

```{r}
###############################useless##############################################
##mtor
#just trying some stuff on a particular example
mtor_cnv=(brain_cancer_cnv[,grep("MTOR", colnames(brain_cancer_cnv))])
mtor_exp=(brain_cancer_exp[,grep("MTOR", colnames(brain_cancer_exp))])
mtor_achilles=(brain_cancer_achilles[,grep("MTOR", colnames(brain_cancer_achilles))])
cmtor <- apply(mtor_cnv, 2, mean) #all values show a deletion of these genes in our cell lines
emtor <- apply(mtor_exp, 2, mean) #however all of them seem to be over expressed
amtor <- apply(mtor_achilles, 2, mean)# more essential (mTOR)
MTOR <- c(cmtor[1],emtor[6],amtor[6])


##top
top_cnv=(brain_cancer_cnv[,grep("TOP", colnames(brain_cancer_cnv))])
top_exp=(brain_cancer_exp[,grep("TOP", colnames(brain_cancer_exp))])
top_achilles=(brain_cancer_achilles[,grep("TOP", colnames(brain_cancer_achilles))])
ctop <- apply(top_cnv, 2, mean) #all values show a deletion of these genes in our cell lines, approx. only 1 allele is deleted
etop <- apply(top_exp, 2, mean) #however very unique expression levels
atop <- apply(top_achilles, 2, mean)#some are more essential (TOP2A,TOP3A) than others
TOP1 <- c(ctop[7],etop[13],atop[5])
TOP2A <- c(ctop[4],etop[3],atop[7])
TOP2B <- c(ctop[10],etop[1],atop[8])

##AURK
aurk_cnv=(brain_cancer_cnv[,grep("AURK", colnames(brain_cancer_cnv))])
aurk_exp=(brain_cancer_exp[,grep("AURK", colnames(brain_cancer_exp))])
aurk_achilles=(brain_cancer_achilles[,grep("AURK", colnames(brain_cancer_achilles))])
caurk <- apply(aurk_cnv, 2, mean) #all values show a deletion of these genes in our cell lines, approx. only 1 allele is deleted
eaurk <- apply(aurk_exp, 2, mean) #however some of them seem to be over expressed: AURKC <0 vs. AURKA/AIP1/B > 6
aaurk <- apply(aurk_achilles, 2, mean)#some are more essential (AURKA, AURKB)
AURKA <- c(caurk[5],eaurk[1],aaurk[1])
AURKB <- c(caurk[3],eaurk[4],aaurk[3])
AURKC <- c(caurk[4],eaurk[2],aaurk[4])

##JAK
jak_cnv=(brain_cancer_cnv[,grep("JAK", colnames(brain_cancer_cnv))])
jak_exp=(brain_cancer_exp[,grep("JAK", colnames(brain_cancer_exp))])
jak_achilles=(brain_cancer_achilles[,grep("JAK", colnames(brain_cancer_achilles))])
cjak <- apply(jak_cnv, 2, mean) #all values show a deletion of these genes in our cell lines, approx. only 1 allele is deleted
ejak <- apply(jak_exp, 2, mean) #seem to be over expressed: JAK2 (2,5), JAK1 (5,7)
ajak <- apply(jak_achilles, 2, mean)#max. is -0,2 (JAK3)
JAK1 <- c(cjak[1],ejak[4],ajak[1])
JAK2 <- c(cjak[7],ejak[1],ajak[2])
JAK3 <- c(cjak[3],ejak[2],ajak[3])

##KIF
kif_cnv=(brain_cancer_cnv[,grep("KIF", colnames(brain_cancer_cnv))])
kif_exp=(brain_cancer_exp[,grep("KIF", colnames(brain_cancer_exp))])
kif_achilles=(brain_cancer_achilles[,grep("KIF", colnames(brain_cancer_achilles))])
ckif <- apply(kif_cnv, 2, mean) 
ekif <- apply(kif_exp, 2, mean) 
akif <- apply(kif_achilles, 2, mean)
KIF11 <- c(ckif[12],ekif[23],akif[1])

##KCNT2
kcnt2_cnv=(brain_cancer_cnv[,grep("KCNT", colnames(brain_cancer_cnv))])
kcnt2_exp=(brain_cancer_exp[,grep("KCNT", colnames(brain_cancer_exp))])
kcnt2_achilles=(brain_cancer_achilles[,grep("KCNT", colnames(brain_cancer_achilles))])
ckcnt <- apply(kcnt2_cnv, 2, mean) 
ekcnt <- apply(kcnt2_exp, 2, mean) 
akcnt <- apply(kcnt2_achilles, 2, mean)
KCNT2 <- c(ckcnt[1],ekcnt[2],akcnt[2])

##PIK
pik_cnv=(brain_cancer_cnv[,grep("PIK", colnames(brain_cancer_cnv))])
pik_exp=(brain_cancer_exp[,grep("PIK", colnames(brain_cancer_exp))])
pik_achilles=(brain_cancer_achilles[,grep("PIK", colnames(brain_cancer_achilles))])
cpik <- apply(pik_cnv, 2, mean) 
epik <- apply(pik_exp, 2, mean) 
apik <- apply(pik_achilles, 2, mean)
PIK3CA <- c(cpik[19],epik[9],apik[6])
PIK3CD <- c(cpik[1],epik[15],apik[8])
PIK3CG <- c(cpik[21],epik[6],apik[9])
PIK3CB <- c(cpik[18],epik[2],apik[7])

##PSM
psm_cnv=(brain_cancer_cnv[,grep("PSM", colnames(brain_cancer_cnv))])
psm_exp=(brain_cancer_exp[,grep("PSM", colnames(brain_cancer_exp))])
psm_achilles=(brain_cancer_achilles[,grep("PSM", colnames(brain_cancer_achilles))])
cpsm <- apply(psm_cnv, 2, mean) 
epsm <- apply(psm_exp, 2, mean) 
apsm <- apply(psm_achilles, 2, mean)
PSMA1 <- c(cpsm[7],epsm[27],apsm[4])
PSMA2 <- c(cpsm[51],epsm[18],apsm[5])
PSMA3 <- c(cpsm[16],epsm[10],apsm[6])
PSMA4 <- c(cpsm[19],epsm[3],apsm[7])
PSMA5 <- c(cpsm[3],epsm[32],apsm[8])
PSMA6 <- c(cpsm[14],epsm[13],apsm[9])
PSMA7 <- c(cpsm[38],epsm[15],apsm[10])
PSMA8 <- c(cpsm[31],epsm[33],apsm[11])
PSMB1 <- c(cpsm[48],epsm[1],apsm[12])
PSMB2 <- c(cpsm[1],epsm[25],apsm[15])
PSMB3 <- c(cpsm[24],epsm[53],apsm[16])
PSMB4 <- c(cpsm[5],epsm[36],apsm[17])
PSMB5 <- c(cpsm[10],epsm[12],apsm[18])
PSMB6 <- c(cpsm[22],epsm[31],apsm[19])
PSMB7 <- c(cpsm[54],epsm[30],apsm[20])
PSMB8 <- c(cpsm[45],epsm[48],apsm[21])
PSMB9 <- c(cpsm[47],epsm[52],apsm[22])
PSMB10 <- c(cpsm[20],epsm[49],apsm[13])
PSMB11 <- c(cpsm[11],epsm[51],apsm[14])

##CDK
cdk_cnv=(brain_cancer_cnv[,grep("CDK", colnames(brain_cancer_cnv))])
cdk_exp=(brain_cancer_exp[,grep("CDK", colnames(brain_cancer_exp))])
cdk_achilles=(brain_cancer_achilles[,grep("CDK", colnames(brain_cancer_achilles))])
ccdk <- apply(cdk_cnv, 2, mean) 
ecdk <- apply(cdk_exp, 2, mean) 
acdk <- apply(cdk_achilles, 2, mean)
CDK1 <- c(ccdk[5],ecdk[37],acdk[1])
CDK2 <- c(ccdk[9],ecdk[17],acdk[12])
CDK4 <- c(ccdk[10],ecdk[23],acdk[17])
CDK5 <- c(ccdk[38],ecdk[33],acdk[18])
CDK6 <- c(ccdk[37],ecdk[11],acdk[24])
CDK7 <- c(ccdk[29],ecdk[22],acdk[25])
CDK8 <- c(ccdk[13],ecdk[21],acdk[26])
CDK9 <- c(ccdk[46],ecdk[24],acdk[27])

##BCL2
bcl_cnv=(brain_cancer_cnv[,grep("BCL", colnames(brain_cancer_cnv))])
bcl_exp=(brain_cancer_exp[,grep("BCL", colnames(brain_cancer_exp))])
bcl_achilles=(brain_cancer_achilles[,grep("BCL", colnames(brain_cancer_achilles))])
cbcl <- apply(bcl_cnv, 2, mean) 
ebcl <- apply(bcl_exp, 2, mean) 
abcl <- apply(bcl_achilles, 2, mean)
BCL2 <- c(cbcl[14],ebcl[20],abcl[4])

##TRPV4
trpv_cnv=(brain_cancer_cnv[,grep("TRPV", colnames(brain_cancer_cnv))])
trpv_exp=(brain_cancer_exp[,grep("TRPV", colnames(brain_cancer_exp))])
trpv_achilles=(brain_cancer_achilles[,grep("TRPV", colnames(brain_cancer_achilles))])
ctrpv <- apply(trpv_cnv, 2, mean) 
etrpv <- apply(trpv_exp, 2, mean) 
atrpv <- apply(trpv_achilles, 2, mean)
TRPV4 <- c(ctrpv[1],etrpv[1],atrpv[4])

##HSP90AA1
hsp_cnv=(brain_cancer_cnv[,grep("HSP", colnames(brain_cancer_cnv))])
hsp_exp=(brain_cancer_exp[,grep("HSP", colnames(brain_cancer_exp))])
hsp_achilles=(brain_cancer_achilles[,grep("HSP", colnames(brain_cancer_achilles))])
chsp <- apply(hsp_cnv, 2, mean) 
ehsp <- apply(hsp_exp, 2, mean) 
ahsp <- apply(hsp_achilles, 2, mean)
HSP90AA1 <- c(chsp[15],ehsp[3],ahsp[3])

##FGF2
fgf_cnv=(brain_cancer_cnv[,grep("FGF", colnames(brain_cancer_cnv))])
fgf_exp=(brain_cancer_exp[,grep("FGF", colnames(brain_cancer_exp))])
fgf_achilles=(brain_cancer_achilles[,grep("FGF", colnames(brain_cancer_achilles))])
cfgf <- apply(fgf_cnv, 2, mean) 
efgf <- apply(fgf_exp, 2, mean) 
afgf <- apply(fgf_achilles, 2, mean)
FGF2<- c(cfgf[27],efgf[22],afgf[10])

##FKBP1A
fkb_cnv=(brain_cancer_cnv[,grep("FKBP1", colnames(brain_cancer_cnv))])
fkb_exp=(brain_cancer_exp[,grep("FKBP1", colnames(brain_cancer_exp))])
fkb_achilles=(brain_cancer_achilles[,grep("FKBP1", colnames(brain_cancer_achilles))])
cfkb <- apply(fkb_cnv, 2, mean) 
efkb <- apply(fkb_exp, 2, mean) 
afkb <- apply(fkb_achilles, 2, mean)
FKBP1A <- c(cfkb[6],efkb[1],afkb[5])

##DHFR
dhf_cnv=(brain_cancer_cnv[,grep("DHF", colnames(brain_cancer_cnv))])
dhf_exp=(brain_cancer_exp[,grep("DHF", colnames(brain_cancer_exp))])
dhf_achilles=(brain_cancer_achilles[,grep("DHF", colnames(brain_cancer_achilles))])
cdhf <- apply(dhf_cnv, 2, mean) 
edhf <- apply(dhf_exp, 2, mean) 
adhf <- apply(dhf_achilles, 2, mean)
DHFR <- c(cdhf[3],edhf[3],adhf[2])

##TYMS
tym_cnv=(brain_cancer_cnv[,grep("TYM", colnames(brain_cancer_cnv))])
tym_exp=(brain_cancer_exp[,grep("TYM", colnames(brain_cancer_exp))])
tym_achilles=(brain_cancer_achilles[,grep("TYM", colnames(brain_cancer_achilles))])
ctym <- apply(tym_cnv, 2, mean) 
etym <- apply(tym_exp, 2, mean) 
atym <- apply(tym_achilles, 2, mean)
TYMS <- c(ctym[2],etym[3],atym[3])

##ATP5A1                                      ????????????????
atp_cnv=(brain_cancer_cnv[,grep("ATP5", colnames(brain_cancer_cnv))])
atp_exp=(brain_cancer_exp[,grep("ATP5", colnames(brain_cancer_exp))])
atp_achilles=(brain_cancer_achilles[,grep("ATP5", colnames(brain_cancer_achilles))])
catp <- apply(atp_cnv, 2, mean) 
eatp <- apply(atp_exp, 2, mean) 
aatp <- apply(atp_achilles, 2, mean)
ATP5A1 <- c(catp[],eatp[],aatp[])

##ATR
atr_cnv=(brain_cancer_cnv[,grep("ATR", colnames(brain_cancer_cnv))])
atr_exp=(brain_cancer_exp[,grep("ATR", colnames(brain_cancer_exp))])
atr_achilles=(brain_cancer_achilles[,grep("ATR", colnames(brain_cancer_achilles))])
catr <- apply(atr_cnv, 2, mean) 
eatr <- apply(atr_exp, 2, mean) 
aatr <- apply(atr_achilles, 2, mean)
ATR <- c(catr[13],eatr[12],aatr[1])

##LDLR
ldl_cnv=(brain_cancer_cnv[,grep("LDLR", colnames(brain_cancer_cnv))])
ldl_exp=(brain_cancer_exp[,grep("LDLR", colnames(brain_cancer_exp))])
ldl_achilles=(brain_cancer_achilles[,grep("LDLR", colnames(brain_cancer_achilles))])
cldl <- apply(ldl_cnv, 2, mean) 
eldl <- apply(ldl_exp, 2, mean) 
aldl <- apply(ldl_achilles, 2, mean)
LDLR <- c(cldl[7],eldl[1],aldl[1])

##CMPK1
cmp_cnv=(brain_cancer_cnv[,grep("CMPK", colnames(brain_cancer_cnv))])
cmp_exp=(brain_cancer_exp[,grep("CMPK", colnames(brain_cancer_exp))])
cmp_achilles=(brain_cancer_achilles[,grep("CMPK", colnames(brain_cancer_achilles))])
ccmp <- apply(cmp_cnv, 2, mean) 
ecmp <- apply(cmp_exp, 2, mean) 
acmp <- apply(cmp_achilles, 2, mean)
CMPK1 <- c(ccmp[1],ecmp[2],acmp[1])

##RRM
rrm_cnv=(brain_cancer_cnv[,grep("RRM", colnames(brain_cancer_cnv))])
rrm_exp=(brain_cancer_exp[,grep("RRM", colnames(brain_cancer_exp))])
rrm_achilles=(brain_cancer_achilles[,grep("RRM", colnames(brain_cancer_achilles))])
crrm <- apply(rrm_cnv, 2, mean) 
errm <- apply(rrm_exp, 2, mean) 
arrm <- apply(rrm_achilles, 2, mean)
RRM1 <- c(crrm[2],errm[4],arrm[1])
RRM2 <- c(crrm[7],errm[6],arrm[2])
RRM2B <- c(crrm[9],errm[1],arrm[3])

##MMP
mmp_cnv=(brain_cancer_cnv[,grep("MMP", colnames(brain_cancer_cnv))])
mmp_exp=(brain_cancer_exp[,grep("MMP", colnames(brain_cancer_exp))])
mmp_achilles=(brain_cancer_achilles[,grep("MMP", colnames(brain_cancer_achilles))])
cmmp <- apply(mmp_cnv, 2, mean) 
emmp <- apply(mmp_exp, 2, mean) 
ammp <- apply(mmp_achilles, 2, mean)
MMP1 <- c(cmmp[11],emmp[23],ammp[3])
MMP2 <- c(cmmp[24],emmp[2],ammp[12])
MMP3 <- c(cmmp[12],emmp[15],ammp[21])
MMP7 <- c(cmmp[6],emmp[10],ammp[22])
MMP8 <- c(cmmp[9],emmp[6],ammp[23])
MMP9 <- c(cmmp[30],emmp[4],ammp[24])
MMP10 <- c(cmmp[10],emmp[19],ammp[4])
MMP11 <- c(cmmp[31],emmp[3],ammp[5])
MMP12 <- c(cmmp[13],emmp[25],ammp[])#?
MMP13 <- c(cmmp[14],emmp[13],ammp[6])
MMP14 <- c(cmmp[17],emmp[18],ammp[7])
MMP15 <- c(cmmp[25],emmp[5],ammp[8])
MMP16 <- c(cmmp[33],emmp[17],ammp[9])
MMP17 <- c(cmmp[16],emmp[24],ammp[10])
MMP19 <- c(cmmp[15],emmp[7],ammp[11])
MMP20 <- c(cmmp[7],emmp[11],ammp[13])
MMP21 <- c(cmmp[3],emmp[16],ammp[14])
MMP23A <- c(cmmp[2],emmp[],ammp[])#????????????????????????????
MMP24 <- c(cmmp[28],emmp[8],ammp[16])
MMP25 <- c(cmmp[21],emmp[1],ammp[17])
MMP26 <- c(cmmp[4],emmp[20],ammp[18])
MMP27 <- c(cmmp[8],emmp[12],ammp[19])
MMP28 <- c(cmmp[26],emmp[26],ammp[20])

##AKT
akt_cnv=(brain_cancer_cnv[,grep("AKT", colnames(brain_cancer_cnv))])
akt_exp=(brain_cancer_exp[,grep("AKT", colnames(brain_cancer_exp))])
akt_achilles=(brain_cancer_achilles[,grep("AKT", colnames(brain_cancer_achilles))])
cakt <- apply(akt_cnv, 2, mean) 
eakt <- apply(akt_exp, 2, mean) 
aakt <- apply(akt_achilles, 2, mean)
AKT1 <- c(cakt[2],eakt[3],aakt[1])
AKT2 <- c(cakt[4],eakt[1],aakt[3])
AKT3 <- c(cakt[1],eakt[2],aakt[4])

##RARG
rar_cnv=(brain_cancer_cnv[,grep("RARG", colnames(brain_cancer_cnv))])
rar_exp=(brain_cancer_exp[,grep("RARG", colnames(brain_cancer_exp))])
rar_achilles=(brain_cancer_achilles[,grep("RARG", colnames(brain_cancer_achilles))])
crar <- apply(rar_cnv, 2, mean) 
erar <- apply(rar_exp, 2, mean) 
arar <- apply(rar_achilles, 2, mean)
RARG <- c(crar[1],erar[1],arar[1])

##RELA
rel_cnv=(brain_cancer_cnv[,grep("REL", colnames(brain_cancer_cnv))])
rel_exp=(brain_cancer_exp[,grep("REL", colnames(brain_cancer_exp))])
rel_achilles=(brain_cancer_achilles[,grep("REL", colnames(brain_cancer_achilles))])
crel <- apply(rel_cnv, 2, mean) 
erel <- apply(rel_exp, 2, mean) 
arel <- apply(rel_achilles, 2, mean)
RELA <- c(crel[4],erel[13],arel[14])

##ABL
abl_cnv=(brain_cancer_cnv[,grep("ABL", colnames(brain_cancer_cnv))])
abl_exp=(brain_cancer_exp[,grep("ABL", colnames(brain_cancer_exp))])
abl_achilles=(brain_cancer_achilles[,grep("ABL", colnames(brain_cancer_achilles))])
cabl <- apply(abl_cnv, 2, mean) 
eabl <- apply(abl_exp, 2, mean) 
aabl <- apply(abl_achilles, 2, mean)
ABL1 <- c(cabl[11],eabl[2],aabl[1])
ABL2 <- c(cabl[1],eabl[5],aabl[2])

##BLK
blk_cnv=(brain_cancer_cnv[,grep("BL", colnames(brain_cancer_cnv))])
blk_exp=(brain_cancer_exp[,grep("BL", colnames(brain_cancer_exp))])
blk_achilles=(brain_cancer_achilles[,grep("BL", colnames(brain_cancer_achilles))])
cblk <- apply(blk_cnv, 2, mean) 
eblk <- apply(blk_exp, 2, mean) 
ablk <- apply(blk_achilles, 2, mean)
BLK <- c(cblk[102],eblk[40],ablk[14])

##EPHA2
eph_cnv=(brain_cancer_cnv[,grep("EPHA", colnames(brain_cancer_cnv))])
eph_exp=(brain_cancer_exp[,grep("EPHA", colnames(brain_cancer_exp))])
eph_achilles=(brain_cancer_achilles[,grep("EPHA", colnames(brain_cancer_achilles))])
ceph <- apply(eph_cnv, 2, mean) 
eeph <- apply(eph_exp, 2, mean) 
aeph <- apply(eph_achilles, 2, mean)
EPHA2 <- c(ceph[1],eeph[6],aeph[3])

##FGR
fgr_cnv=(brain_cancer_cnv[,grep("FG", colnames(brain_cancer_cnv))])
fgr_exp=(brain_cancer_exp[,grep("FG", colnames(brain_cancer_exp))])
fgr_achilles=(brain_cancer_achilles[,grep("FG", colnames(brain_cancer_achilles))])
cfgr <- apply(fgr_cnv, 2, mean) 
efgr <- apply(fgr_exp, 2, mean) 
afgr <- apply(fgr_achilles, 2, mean)
FGR <- c(cfgr[1],efgr[1],afgr[54])

##FRK
frk_cnv=(brain_cancer_cnv[,grep("FRK", colnames(brain_cancer_cnv))])
frk_exp=(brain_cancer_exp[,grep("FRK", colnames(brain_cancer_exp))])
frk_achilles=(brain_cancer_achilles[,grep("FRK", colnames(brain_cancer_achilles))])
cfrk <- apply(frk_cnv, 2, mean) 
efrk <- apply(frk_exp, 2, mean) 
afrk <- apply(frk_achilles, 2, mean)
FRK <- c(cfrk[2],efrk[1],afrk[1])

##FYN
fyn_cnv=(brain_cancer_cnv[,grep("FY", colnames(brain_cancer_cnv))])
fyn_exp=(brain_cancer_exp[,grep("FY", colnames(brain_cancer_exp))])
fyn_achilles=(brain_cancer_achilles[,grep("FY", colnames(brain_cancer_achilles))])
cfyn <- apply(fyn_cnv, 2, mean) 
efyn <- apply(fyn_exp, 2, mean) 
afyn <- apply(fyn_achilles, 2, mean)
FYN <- c(cfyn[32],efyn[2],afyn[5])

##HCK
hck_cnv=(brain_cancer_cnv[,grep("HC", colnames(brain_cancer_cnv))])
hck_exp=(brain_cancer_exp[,grep("HC", colnames(brain_cancer_exp))])
hck_achilles=(brain_cancer_achilles[,grep("HC", colnames(brain_cancer_achilles))])
chck <- apply(hck_cnv, 2, mean) 
ehck <- apply(hck_exp, 2, mean) 
ahck <- apply(hck_achilles, 2, mean)
HCK <- c(chck[65],ehck[9],ahck[26])

##KIT
kit_cnv=(brain_cancer_cnv[,grep("KIT", colnames(brain_cancer_cnv))])
kit_exp=(brain_cancer_exp[,grep("KIT", colnames(brain_cancer_exp))])
kit_achilles=(brain_cancer_achilles[,grep("KIT", colnames(brain_cancer_achilles))])
ckit <- apply(kit_cnv, 2, mean) 
ekit <- apply(kit_exp, 2, mean) 
akit <- apply(kit_achilles, 2, mean)
KIT <- c(ckit[2],ekit[2],akit[1])

##LCK
lck_cnv=(brain_cancer_cnv[,grep("LC", colnames(brain_cancer_cnv))])
lck_exp=(brain_cancer_exp[,grep("LC", colnames(brain_cancer_exp))])
lck_achilles=(brain_cancer_achilles[,grep("LC", colnames(brain_cancer_achilles))])
clck <- apply(lck_cnv, 2, mean) 
elck <- apply(lck_exp, 2, mean) 
alck <- apply(lck_achilles, 2, mean)
LCK <- c(clck[16],elck[421],alck[64])

##LYN
lyn_cnv=(brain_cancer_cnv[,grep("LY", colnames(brain_cancer_cnv))])
lyn_exp=(brain_cancer_exp[,grep("LY", colnames(brain_cancer_exp))])
lyn_achilles=(brain_cancer_achilles[,grep("LY", colnames(brain_cancer_achilles))])
clyn <- apply(lyn_cnv, 2, mean) 
elyn <- apply(lyn_exp, 2, mean) 
alyn <- apply(lyn_achilles, 2, mean)
LYN <- c(clyn[77],elyn[73],alyn[31])

##PDGFRB
pdg_cnv=(brain_cancer_cnv[,grep("PDGF", colnames(brain_cancer_cnv))])
pdg_exp=(brain_cancer_exp[,grep("PDGF", colnames(brain_cancer_exp))])
pdg_achilles=(brain_cancer_achilles[,grep("PDGF", colnames(brain_cancer_achilles))])
cpdg <- apply(pdg_cnv, 2, mean) 
epdg <- apply(pdg_exp, 2, mean) 
apdg <- apply(pdg_achilles, 2, mean)
PDGFRB <- c(cpdg[5],epdg[3],apdg[6])

##SRC
src_cnv=(brain_cancer_cnv[,grep("SRC", colnames(brain_cancer_cnv))])
src_exp=(brain_cancer_exp[,grep("SRC", colnames(brain_cancer_exp))])
src_achilles=(brain_cancer_achilles[,grep("SRC", colnames(brain_cancer_achilles))])
csrc <- apply(src_cnv, 2, mean) 
esrc <- apply(src_exp, 2, mean) 
asrc <- apply(src_achilles, 2, mean)
SRC <- c(csrc[5],esrc[5],asrc[4])

##SRMS
srm_cnv=(brain_cancer_cnv[,grep("SRM", colnames(brain_cancer_cnv))])
srm_exp=(brain_cancer_exp[,grep("SRM", colnames(brain_cancer_exp))])
srm_achilles=(brain_cancer_achilles[,grep("SRM", colnames(brain_cancer_achilles))])
csrm <- apply(srm_cnv, 2, mean) 
esrm <- apply(srm_exp, 2, mean) 
asrm <- apply(srm_achilles, 2, mean)
SRMS <- c(csrm[2],esrm[2],asrm[2])

##STAT5B
sta_cnv=(brain_cancer_cnv[,grep("STAT", colnames(brain_cancer_cnv))])
sta_exp=(brain_cancer_exp[,grep("STAT", colnames(brain_cancer_exp))])
sta_achilles=(brain_cancer_achilles[,grep("STAT", colnames(brain_cancer_achilles))])
csta <- apply(sta_cnv, 2, mean) 
esta <- apply(sta_exp, 2, mean) 
asta <- apply(sta_achilles, 2, mean)
STAT5B <- c(csta[3],esta[8],asta[6])

##YES1
yes_cnv=(brain_cancer_cnv[,grep("YE", colnames(brain_cancer_cnv))])
yes_exp=(brain_cancer_exp[,grep("YE", colnames(brain_cancer_exp))])
yes_achilles=(brain_cancer_achilles[,grep("YE", colnames(brain_cancer_achilles))])
cyes <- apply(yes_cnv, 2, mean) 
eyes <- apply(yes_exp, 2, mean) 
ayes <- apply(yes_achilles, 2, mean)
YES1 <- c(cyes[5],eyes[8],ayes[8])

##IMPDH
imp_cnv=(brain_cancer_cnv[,grep("IMPDH", colnames(brain_cancer_cnv))])
imp_exp=(brain_cancer_exp[,grep("IMPDH", colnames(brain_cancer_exp))])
imp_achilles=(brain_cancer_achilles[,grep("IMPDH", colnames(brain_cancer_achilles))])
cimp <- apply(imp_cnv, 2, mean) 
eimp <- apply(imp_exp, 2, mean) 
aimp <- apply(imp_achilles, 2, mean)
IMPDH1 <- c(cimp[2],eimp[1],aimp[1])
IMPDH2 <- c(cimp[1],eimp[2],aimp[2])

##MELK
mel_cnv=(brain_cancer_cnv[,grep("MEL", colnames(brain_cancer_cnv))])
mel_exp=(brain_cancer_exp[,grep("MEL", colnames(brain_cancer_exp))])
mel_achilles=(brain_cancer_achilles[,grep("MEL", colnames(brain_cancer_achilles))])
cmel <- apply(mel_cnv, 2, mean) 
emel <- apply(mel_exp, 2, mean) 
amel <- apply(mel_achilles, 2, mean)
MELK <- c(cmel[7],emel[6],amel[2])

##PLK1
plk_cnv=(brain_cancer_cnv[,grep("PLK", colnames(brain_cancer_cnv))])
plk_exp=(brain_cancer_exp[,grep("PLK", colnames(brain_cancer_exp))])
plk_achilles=(brain_cancer_achilles[,grep("PLK", colnames(brain_cancer_achilles))])
cplk <- apply(plk_cnv, 2, mean) 
eplk <- apply(plk_exp, 2, mean) 
aplk <- apply(plk_achilles, 2, mean)
PLK1 <- c(cplk[2],eplk[3],aplk[2])

##NOS
nos_cnv=(brain_cancer_cnv[,grep("NOS", colnames(brain_cancer_cnv))])
nos_exp=(brain_cancer_exp[,grep("NOS", colnames(brain_cancer_exp))])
nos_achilles=(brain_cancer_achilles[,grep("NOS", colnames(brain_cancer_achilles))])
cnos <- apply(nos_cnv, 2, mean) 
enos <- apply(nos_exp, 2, mean) 
anos <- apply(nos_achilles, 2, mean)
NOS1 <- c(cnos[5],enos[3],anos[7])
NOS2 <- c(cnos[9],enos[1],anos[9])

##ACVR1C
acv_cnv=(brain_cancer_cnv[,grep("ACVR", colnames(brain_cancer_cnv))])
acv_exp=(brain_cancer_exp[,grep("ACVR", colnames(brain_cancer_exp))])
acv_achilles=(brain_cancer_achilles[,grep("ACVR", colnames(brain_cancer_achilles))])
cacv <- apply(acv_cnv, 2, mean) 
eacv <- apply(acv_exp, 2, mean) 
aacv <- apply(acv_achilles, 2, mean)
ACVR1C <- c(cacv[4],eacv[4],aacv[3])

##TGFBR1
tgf_cnv=(brain_cancer_cnv[,grep("TGFBR", colnames(brain_cancer_cnv))])
tgf_exp=(brain_cancer_exp[,grep("TGFBR", colnames(brain_cancer_exp))])
tgf_achilles=(brain_cancer_achilles[,grep("TGFBR", colnames(brain_cancer_achilles))])
ctgf <- apply(tgf_cnv, 2, mean) 
etgf <- apply(tgf_exp, 2, mean) 
atgf <- apply(tgf_achilles, 2, mean)
TGFBR1 <- c(ctgf[5],etgf[2],atgf[1])

##BIRC5
bir_cnv=(brain_cancer_cnv[,grep("BIRC", colnames(brain_cancer_cnv))])
bir_exp=(brain_cancer_exp[,grep("BIRC", colnames(brain_cancer_exp))])
bir_achilles=(brain_cancer_achilles[,grep("BIRC", colnames(brain_cancer_achilles))])
cbir <- apply(bir_cnv, 2, mean) 
ebir <- apply(bir_exp, 2, mean) 
abir <- apply(bir_achilles, 2, mean)
BIRC5 <- c(cbir[3],ebir[2],abir[3])

##TUBB
tub_cnv=(brain_cancer_cnv[,grep("TUBB", colnames(brain_cancer_cnv))])
tub_exp=(brain_cancer_exp[,grep("TUBB", colnames(brain_cancer_exp))])
tub_achilles=(brain_cancer_achilles[,grep("TUBB", colnames(brain_cancer_achilles))])
ctub <- apply(tub_cnv, 2, mean) 
etub <- apply(tub_exp, 2, mean) 
atub <- apply(tub_achilles, 2, mean)
TUBB <- c(ctub[9],etub[8],atub[1])

##EGFR
egf_cnv=(brain_cancer_cnv[,grep("EGF", colnames(brain_cancer_cnv))])
egf_exp=(brain_cancer_exp[,grep("EGF", colnames(brain_cancer_exp))])
egf_achilles=(brain_cancer_achilles[,grep("EGF", colnames(brain_cancer_achilles))])
cegf <- apply(egf_cnv, 2, mean) 
eegf <- apply(egf_exp, 2, mean) 
aegf <- apply(egf_achilles, 2, mean)
EGFR <- c(cegf[16],eegf[7],aegf[6])

##PYGM
pyg_cnv=(brain_cancer_cnv[,grep("PYG", colnames(brain_cancer_cnv))])
pyg_exp=(brain_cancer_exp[,grep("PYG", colnames(brain_cancer_exp))])
pyg_achilles=(brain_cancer_achilles[,grep("PYG", colnames(brain_cancer_achilles))])
cpyg <- apply(pyg_cnv, 2, mean) 
epyg <- apply(pyg_exp, 2, mean) 
apyg <- apply(pyg_achilles, 2, mean)
PYGM <- c(cpyg[2],epyg[1],apyg[3])

##ANPEP
anp_cnv=(brain_cancer_cnv[,grep("ANP", colnames(brain_cancer_cnv))])
anp_exp=(brain_cancer_exp[,grep("ANP", colnames(brain_cancer_exp))])
anp_achilles=(brain_cancer_achilles[,grep("ANP", colnames(brain_cancer_achilles))])
canp <- apply(anp_cnv, 2, mean) 
eanp <- apply(anp_exp, 2, mean) 
aanp <- apply(anp_achilles, 2, mean)
ANPEP <- c(canp[7],eanp[6],aanp[5])

##LAP3
lap_cnv=(brain_cancer_cnv[,grep("LAP", colnames(brain_cancer_cnv))])
lap_exp=(brain_cancer_exp[,grep("LAP", colnames(brain_cancer_exp))])
lap_achilles=(brain_cancer_achilles[,grep("LAP", colnames(brain_cancer_achilles))])
clap <- apply(lap_cnv, 2, mean) 
elap <- apply(lap_exp, 2, mean) 
alap <- apply(lap_achilles, 2, mean)
LAP3 <- c(clap[6],elap[1],alap[2])

##NPEPPS
npe_cnv=(brain_cancer_cnv[,grep("NPEP", colnames(brain_cancer_cnv))])
npe_exp=(brain_cancer_exp[,grep("NPEP", colnames(brain_cancer_exp))])
npe_achilles=(brain_cancer_achilles[,grep("NPEP", colnames(brain_cancer_achilles))])
cnpe <- apply(npe_cnv, 2, mean) 
enpe <- apply(npe_exp, 2, mean) 
anpe <- apply(npe_achilles, 2, mean)
NPEPPS <- c(cnpe[4],enpe[6],anpe[6])

##ADA
ada_cnv=(brain_cancer_cnv[,grep("ADA", colnames(brain_cancer_cnv))])
ada_exp=(brain_cancer_exp[,grep("ADA", colnames(brain_cancer_exp))])
ada_achilles=(brain_cancer_achilles[,grep("ADA", colnames(brain_cancer_achilles))])
cada <- apply(ada_cnv, 2, mean) 
eada <- apply(ada_exp, 2, mean) 
aada <- apply(ada_achilles, 2, mean)
ADA <- c(cada[45],eada[61],aada[6])

##PNP
pnp_cnv=(brain_cancer_cnv[,grep("PNP", colnames(brain_cancer_cnv))])
pnp_exp=(brain_cancer_exp[,grep("PNP", colnames(brain_cancer_exp))])
pnp_achilles=(brain_cancer_achilles[,grep("PNP", colnames(brain_cancer_achilles))])
cpnp <- apply(pnp_cnv, 2, mean) 
epnp <- apply(pnp_exp, 2, mean) 
apnp <- apply(pnp_achilles, 2, mean)
PNP <- c(cpnp[4],epnp[14],apnp[1])

##POL
pol_cnv=(brain_cancer_cnv[,grep("POL", colnames(brain_cancer_cnv))])
pol_exp=(brain_cancer_exp[,grep("POL", colnames(brain_cancer_exp))])
pol_achilles=(brain_cancer_achilles[,grep("POL", colnames(brain_cancer_achilles))])
cpol <- apply(pol_cnv, 2, mean) 
epol <- apply(pol_exp, 2, mean) 
apol <- apply(pol_achilles, 2, mean)
POLA1 <- c(cpol[66],epol[23],apol[10])
POLE <- c(cpol[12],epol[57],apol[18])
POLE2 <- c(cpol[15],epol[21],apol[19])
POLE3 <- c(cpol[65],epol[45],apol[20])
POLE4 <- c(cpol[31],epol[29],apol[21])

##PARP2
par_cnv=(brain_cancer_cnv[,grep("PARP", colnames(brain_cancer_cnv))])
par_exp=(brain_cancer_exp[,grep("PARP", colnames(brain_cancer_exp))])
par_achilles=(brain_cancer_achilles[,grep("PARP", colnames(brain_cancer_achilles))])
cpar <- apply(par_cnv, 2, mean) 
epar <- apply(par_exp, 2, mean) 
apar <- apply(par_achilles, 2, mean)
PARP2 <- c(cpar[6],epar[5],apar[8])
```


```{r}
#set all means of cnv, exp and achilles in matrix
mmm <- rbind(MTOR,TOP1,TOP2A,TOP2B,AURKA,AURKB,AURKC,JAK1,JAK2,JAK3,KIF11,KCNT2,PIK3CA,PIK3CB,PIK3CD,PIK3CG,PSMA1,PSMA2,PSMA3,PSMA4,PSMA5,PSMA6,PSMA7,PSMA8,PSMB1,PSMB2,PSMB3,PSMB4,PSMB5,PSMB6,PSMB7,PSMB8,PSMB9,PSMB10,PSMB11,CDK1,CDK2,CDK4,CDK5,CDK6,CDK7,CDK8,CDK9,BCL2,TRPV4,HSP90AA1,FGF2,FKBP1A,DHFR,TYMS,ATR,LDLR,CMPK1,RRM1,RRM2,RRM2B,MMP1,MMP2,MMP3,MMP7,MMP8,MMP9,MMP10,MMP11,MMP13,MMP14,MMP15,MMP16,MMP17,MMP19,MMP20,MMP21,MMP24,MMP25,MMP26,MMP27,MMP28,AKT1,AKT2,AKT3,RARG,RELA,ABL1,ABL2,BLK,EPHA2,FGR,FRK,FYN,HCK,KIT,LCK,LYN,PDGFRB,SRC,SRMS,STAT5B,YES1,IMPDH1,IMPDH2,MELK,PLK1,NOS1,NOS2,ACVR1C,TGFBR1,BIRC5,TUBB,EGFR,PYGM,ANPEP,LAP3,NPEPPS,ADA,PNP,POLA1,POLE,POLE2,POLE3,POLE4,PARP2)

colnam <- c("cnv mean","exp mean","achilles mean")
colnames(mmm) <- colnam
#Maximum of cnv mean, exp mean and minimum (more negative more essential) achilles mean
max(mmm[,1])
max(mmm[,2])
min(mmm[,3])
```


```{r}
# Question 2: What are the targets of the effective drugs?

##What are the targets of the effective drugs?


brain_cancer_treat=prism.treat[colnames(brain_cancer),] #selecting only the drugs we previously filtered out from brain_cancer
brain_cancer_treat=brain_cancer_treat[!duplicated(brain_cancer_treat$broad_id),]
rownames(brain_cancer_treat)<-brain_cancer_treat[,1]
effective_bc_treat=brain_cancer_treat[effective_in_all_doses,-c(1,2,3,4,10)]
#summary(effective_bc_treat$target)
#summary(effective_bc_treat$moa)
#expression of these genes would be interesting to look into our cell lines
```

#Target of effective drugs
```{r}
brain_cancer.effective_targets <- effective_bc_treat$target
which(is.na(effective_bc_treat$target))
## da fehlen halt bei diesen 8 Drogen die targets, die könnte man sonst ggf. rausnehmen 
```


##Selecting brain cancer specific cell lines in all data sets
```{r}
brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") #identify the brain cancer cell lines' names
names=brain_cancer_cl[,1] #saving the names as names
#select wanted cell lines from all datasets
brain_cancer=prism[names,]
brain_cancer_achilles=prism.achilles[names,]
## hier ist irgendwas schief gegangen, ab da sind super viele NA's drin 
brain_cancer_cnv=prism.cnv[names,]
brain_cancer_exp=prism.exp[names,]
#brain_cancer_snv=prism.snv[names,] #does not make sense no cell line info, other type of filtering required
#brain_cancer_treat=prism.treat[names,] #does not make sense no cell line info, other type of filtering required
```

# Selecting achilles targets
```{r}
#brain_cancer_achilles.mean <- apply(brain_cancer_achilles,2,mean)
#hist(brain_cancer_achilles.mean, breaks = 200)
#brain_cancer_achilles.var <- apply(brain_cancer_achilles,2,var)#variance
#hist(brain_cancer_achilles.var, breaks = 200, xlim = c(0,0.4))
#brain_cancer_achilles.relevant <- colnames(brain_cancer_achilles)[brain_cancer_achilles.mean < -0.2]
```

# Compare drug and achilles targets
```{r}
brain_cancer_treat=prism.treat[colnames(brain_cancer),] #selecting only the drugs we previously filtered out from brain_cancer
brain_cancer_treat=brain_cancer_treat[!duplicated(brain_cancer_treat$broad_id),]
rownames(brain_cancer_treat)<-brain_cancer_treat[,1]
effective_bc_treat=brain_cancer_treat[effective_in_all_doses,-c(1,2,3,4,10)]
#summary(effective_bc_treat$target)
#summary(effective_bc_treat$moa)
#expression of these genes would be interesting to look into our cell lines

brain_cancer.effective_targets <- effective_bc_treat$target
which(is.na(effective_bc_treat$target))

brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") #identify the brain cancer cell lines' names
names=brain_cancer_cl[,1] #saving the names as names
#select wanted cell lines from all datasets
brain_cancer=prism[names,]
brain_cancer_achilles=prism.achilles[names,]
## hier ist irgendwas schief gegangen, ab da sind super viele NA's drin 
brain_cancer_cnv=prism.cnv[names,]
brain_cancer_exp=prism.exp[names,]
#brain_cancer_snv=prism.snv[names,] #does not make sense no cell line info, other type of filtering required
#brain_cancer_treat=prism.treat[names,] #does not make sense no cell line info, other type of filtering required

brain_cancer.effective_targets <- as.character(brain_cancer.effective_targets) # convert to character
x <- 1
brain_cancer.effective_targets.solo <- list() #create list
for(i in brain_cancer.effective_targets) #split multiple targets
  {
  
 brain_cancer.effective_targets.solo[[x]] <- strsplit(i, split = ", ")
 x <- x+1
 }
brain_cancer.effective_targets.solo <- unlist(brain_cancer.effective_targets.solo) #list to vector
brain_cancer.effective_targets.solo <- unique(brain_cancer.effective_targets.solo) # remove multiple targets
brain_cancer.effective_targets.solo <- brain_cancer.effective_targets.solo[-1] # remove NA
# achilles targets
brain_cancer_achilles.targets <- brain_cancer_achilles[,which(colnames(brain_cancer_achilles) %in% brain_cancer.effective_targets.solo)] # selecting target genes in achilles
brain_cancer.effective_targets.solo[-which(brain_cancer.effective_targets.solo %in% colnames(brain_cancer_achilles))] # missing targets
# exp targets
brain_cancer_exp.targets <- brain_cancer_exp[,which(colnames(brain_cancer_exp) %in% brain_cancer.effective_targets.solo)] # selecting target genes in exp
brain_cancer.effective_targets.solo[-which(brain_cancer.effective_targets.solo %in% colnames(brain_cancer_exp))] # missing targets
# cnv targets
brain_cancer_cnv.targets <- brain_cancer_cnv[,which(colnames(brain_cancer_cnv) %in% brain_cancer.effective_targets.solo)] # selecting target genes in cnvVie
brain_cancer.effective_targets.solo[-which(brain_cancer.effective_targets.solo %in% colnames(brain_cancer_cnv))] # missing targets


```

# Compare drug and achilles targets
```{r}
brain_cancer.effective_targets <- as.character(brain_cancer.effective_targets) # convert to character

x <- 1
brain_cancer.effective_targets.solo <- list() #create list

for(i in brain_cancer.effective_targets) #split multiple targets
  {
  
 brain_cancer.effective_targets.solo[[x]] <- strsplit(i, split = ", ")
 x <- x+1
 }
brain_cancer.effective_targets.solo <- unlist(brain_cancer.effective_targets.solo) #list to vector

brain_cancer.effective_targets.solo <- unique(brain_cancer.effective_targets.solo) # remove multiple targets

brain_cancer.effective_targets.solo <- brain_cancer.effective_targets.solo[-1] # remove NA

# achilles targets
brain_cancer_achilles.targets <- brain_cancer_achilles[,which(colnames(brain_cancer_achilles) %in% brain_cancer.effective_targets.solo)] # selecting target genes in achilles
brain_cancer.effective_targets.solo[-which(brain_cancer.effective_targets.solo %in% colnames(brain_cancer_achilles))] # missing targets

hist(unlist(brain_cancer_achilles), breaks = 50)
hist(unlist(brain_cancer_achilles.targets), breaks = 50)
summary(unlist(brain_cancer_achilles))
summary(unlist(brain_cancer_achilles.targets))
qqplot(unlist(brain_cancer_achilles), unlist(brain_cancer_achilles.targets))
abline(a = 0, b = 1)

# achilles variance analysis
brain_cancer_achilles.var <- apply(brain_cancer_achilles,2,function(x){var(x, na.rm = TRUE)})#variance
brain_cancer_achilles.targets.var <- apply(brain_cancer_achilles.targets,2, function(x){var(x, na.rm = TRUE)})
hist(unlist(brain_cancer_achilles.var), breaks = 500, xlim = c(0,0.2))
hist(unlist(brain_cancer_achilles.targets.var), breaks = 40, xlim = c(0,0.2))
qqplot(unlist(brain_cancer_achilles.var), unlist(brain_cancer_achilles.targets.var), xlim = c(0,0.2))
abline(a = 0, b = 1)


# exp targets
brain_cancer_exp.targets <- brain_cancer_exp[,which(colnames(brain_cancer_exp) %in% brain_cancer.effective_targets.solo)] # selecting target genes in exp
brain_cancer.effective_targets.solo[-which(brain_cancer.effective_targets.solo %in% colnames(brain_cancer_exp))] # missing targets

hist(unlist(brain_cancer_exp), breaks = 50)
hist(unlist(brain_cancer_exp.targets), breaks = 50)
summary(unlist(brain_cancer_exp))
summary(unlist(brain_cancer_exp.targets))
qqplot(unlist(brain_cancer_exp), unlist(brain_cancer_exp.targets))
abline(a = 0, b = 1)

# exp variance analysis
brain_cancer_exp.var <- apply(brain_cancer_exp,2,function(x){var(x, na.rm = TRUE)})#variance
brain_cancer_exp.targets.var <- apply(brain_cancer_exp.targets,2, function(x){var(x, na.rm = TRUE)})
hist(unlist(brain_cancer_exp.var), breaks = 500)
hist(unlist(brain_cancer_exp.targets.var), breaks = 40)
qqplot(unlist(brain_cancer_exp.var), unlist(brain_cancer_exp.targets.var), xlim = c(0,8))
abline(a = 0, b = 1)


# cnv targets
brain_cancer_cnv.targets <- brain_cancer_cnv[,which(colnames(brain_cancer_cnv) %in% brain_cancer.effective_targets.solo)] # selecting target genes in cnvVie
brain_cancer.effective_targets.solo[-which(brain_cancer.effective_targets.solo %in% colnames(brain_cancer_cnv))] # missing targets

hist(unlist(brain_cancer_cnv), breaks = 50)
hist(unlist(brain_cancer_cnv.targets), breaks = 50)
summary(unlist(brain_cancer_cnv))
summary(unlist(brain_cancer_cnv.targets))
qqplot(unlist(brain_cancer_cnv), unlist(brain_cancer_cnv.targets))
abline(a = 0, b = 1)

# cnv variance analysis
brain_cancer_cnv.var <- apply(brain_cancer_cnv,2,function(x){var(x, na.rm = TRUE)})#variance
brain_cancer_cnv.targets.var <- apply(brain_cancer_cnv.targets,2, function(x){var(x, na.rm = TRUE)})
hist(unlist(brain_cancer_cnv.var), breaks = 500)
hist(unlist(brain_cancer_cnv.targets.var), breaks = 40)
qqplot(unlist(brain_cancer_cnv.var), unlist(brain_cancer_cnv.targets.var), xlim = c(0,0.1), ylim = c(0,0.1))
abline(a = 0, b = 1)
```

Caution: This is a case for the multiplicity problem. While testing thousands of hypotheses simultaneously, the chance of getting false positives is increased. There are two possibilities of adjustment.

First: The Bonferroni correction: Correction of the FWER (family-wise error rate) so the probability of getting at least one false positive during multiple testing. In this way, the significance level alpha is adjustet to alpha/n.

Second: The Benjamini and Hochberg correction. This is a method to control the FDR (false discovery rate) so the proportion of false positives among the significant results. This method is more recommended for gene expression analysis.

https://www.reneshbedre.com/blog/multiple-hypothesis-testing-corrections.html

##Wilcoxon test: compare all gene from 34 brain cancer cell lines with the same gene from all other tumors
## Achilles
```{r}
##### pseudo value
#min_bcaT <- vector(length = 18119)
#for(i in 1:18119){
#  min_bcaT[i] <- min(brain_cancer_achilles_T[i,][which(brain_cancer_achilles_T[i,]>0)])
#}
#min(min_bcaT)

#min_bcatT <- vector(length = 121)
#for(i in 1:121){
#  min_bcatT[i]  <- min(brain_cancer_achilles.targets_T[i,][which(brain_cancer_achilles.targets_T[i,]>0)])
#}
#pseudo_value_a <- min(min_bcatT)
# pseudo_value_a = 0.0004840532
```

```{r}
# Removal of lines with NA from new dataframe
brain_cancer_achilles_T <- na.omit(brain_cancer_achilles)
brain_cancer_achilles.targets_T <- na.omit(brain_cancer_achilles.targets)

#Z-transformation
#pseudo_value_a <- 0.0004840532
#brain_cancer_achilles_T <- brain_cancer_achilles_T + pseudo_value_a
#brain_cancer_achilles.targets_T <- brain_cancer_achilles.targets_T + pseudo_value_a

# Transposition, now 25 columns and different #row
brain_cancer_achilles_T <- as.data.frame(t(scale(as.matrix(t(brain_cancer_achilles_T)))))
brain_cancer_achilles.targets_T <- as.data.frame(t(scale(as.matrix(t(brain_cancer_achilles.targets_T)))))

```

```{r}
# Wilcoxon test
# H0: The distribution of the effective targets (achilles values) is equal to the one of the achilles dataset
# H1: The distribution of the effective targets = achilles values are "less" than the one of the achilles dataset
wilcox_a <- vector(mode = "list",length=ncol(brain_cancer_achilles.targets_T))
for(i in 1:ncol(brain_cancer_achilles.targets_T)) {         
  wilcox_a[[i]] <- wilcox.test(as.numeric(brain_cancer_achilles_T[,i]),as.numeric(brain_cancer_achilles.targets_T[,i]), alternative = "less", paired = F)
}

wilcox_pvalue_a <- vector(length = ncol(brain_cancer_achilles.targets_T))
for(i in 1:ncol(brain_cancer_achilles.targets_T)) {
  wilcox_pvalue_a[i] <- wilcox_a[[i]]$p.value
}

length(which(wilcox_pvalue_a<=0.05))# 59


```
##Adjustment of multiple testing problem 
```{r}
# Bonferroni correction
wilcox_pvalue_a_bonf <- p.adjust(wilcox_pvalue_a, method = "bonferroni", n = length(wilcox_pvalue_a)) 
b_correction_a <- which(wilcox_pvalue_a_bonf<=0.01)
p_values_a_b <- wilcox_pvalue_a_bonf[b_correction_a]
sig_targets_a <- colnames(brain_cancer_achilles.targets_T)[b_correction_a]

hist(p_values_a_b)
length(p_values_a_b)# 44 targets are from another distribution and less
  
# Benjamini-Hochberg correction
wilcox_pvalue_a_BH <- p.adjust(wilcox_pvalue_a, method = "BH", n = length(wilcox_pvalue_a))
bh_correction_a <- which(wilcox_pvalue_a_BH<=0.01) # 44
```


## Expression
```{r}
##### pseudo value
#min_bceT <- vector(length = 19177)
#for(i in 1:19177){
#  min_bceT[i] <- min(brain_cancer_exp[i,][which(brain_cancer_exp[i,]>0)])
#}
#min(min_bceT)

#min_bcetT <- vector(length = 122)
#for(i in 1:122){
#  min_bcetT[i]  <- min(brain_cancer_exp.targets_T[i,][which(brain_cancer_exp.targets_T[i,]>0)])
#}
#pseudo_value_e <- min(min_bcetT)
# pseudo_value_e = 
```

```{r}
# Removal of lines with NA from new dataframe --> actually there are non
brain_cancer_exp_T <- na.omit(brain_cancer_exp)
brain_cancer_exp.targets_T <- na.omit(brain_cancer_exp.targets)

#Z-transformation
#pseudo_value_e <- 
#brain_cancer_exp_T <- brain_cancer_exp_T + pseudo_value_e
#brain_cancer_exp.targets_T <- brain_cancer_exp.targets_T + pseudo_value_e

# Transposition
brain_cancer_exp_T <- as.data.frame(t(scale(as.matrix(t(brain_cancer_exp_T)))))
brain_cancer_exp.targets_T <- as.data.frame(t(scale(as.matrix(t(brain_cancer_exp.targets_T)))))

```

```{r}
# Wilcoxon test
# H0: The distribution of the effective targets (exp values) is equal to the one of the exp dataset
# H1: The distribution of the effective targets = exp values are "" than the one of the exp dataset
wilcox_e <- vector(mode = "list",length=ncol(brain_cancer_exp.targets_T))
for(i in 1:ncol(brain_cancer_exp.targets_T)) {         
  wilcox_e[[i]] <- wilcox.test(as.numeric(brain_cancer_exp_T[,i]),as.numeric(brain_cancer_exp.targets_T[,i]), alternative = "two.sided", paired = F)
}

wilcox_pvalue_e <- vector(length = ncol(brain_cancer_exp.targets_T))
for(i in 1:ncol(brain_cancer_exp.targets_T)) {
  wilcox_pvalue_e[i] <- wilcox_e[[i]]$p.value
}

length(which(wilcox_pvalue_e<=0.05))# 39 (less) /// 112 (two.sided)


```
##Adjustment of multiple testing problem 
```{r}
# Bonferroni correction
wilcox_pvalue_e_bonf <- p.adjust(wilcox_pvalue_e, method = "bonferroni", n = length(wilcox_pvalue_e)) 
b_correction_e <- which(wilcox_pvalue_e_bonf<=0.01)
p_values_e_b <- wilcox_pvalue_e_bonf[b_correction_e]
sig_targets_e <- colnames(brain_cancer_exp.targets_T)[b_correction_e]

hist(p_values_e_b)
length(p_values_e_b)# 34 (less) /// 98 (two.sided)
  
# Benjamini-Hochberg correction
wilcox_pvalue_e_BH <- p.adjust(wilcox_pvalue_e, method = "BH", n = length(wilcox_pvalue_e))
bh_correction_e <- which(wilcox_pvalue_e_BH<=0.01)
```


## Copy Number
```{r}
##### pseudo value
#min_bccT <- vector(length = 27562)
#for(i in 1:27562){
#  min_bccT[i] <- min(brain_cancer_cnv_T[i,][which(brain_cancer_cnv_T[i,]>0)])
#}
#min(min_bccT)

#min_bcctT <- vector(length = 123)
#for(i in 1:123){
#  min_bcctT[i]  <- min(brain_cancer_cnv.targets_T[i,][which(brain_cancer_cnv.targets_T[i,]>0)])
#}
#pseudo_value_c <- min(min_bcctT)
# pseudo_value_c = 
```

```{r}
# Removal of lines with NA from new dataframe --> actually there are non
brain_cancer_cnv_T <- na.omit(brain_cancer_cnv)
brain_cancer_cnv.targets_T <- na.omit(brain_cancer_cnv.targets)

#Z-transformation
#pseudo_value_c <- 
#brain_cancer_cnv_T <- brain_cancer_cnv_T + pseudo_value_c
#brain_cancer_cnv.targets_T <- brain_cancer_cnv.targets_T + pseudo_value_c

# Transposition
brain_cancer_cnv_T <- as.data.frame(t(scale(as.matrix(t(brain_cancer_cnv_T)))))
brain_cancer_cnv.targets_T <- as.data.frame(t(scale(as.matrix(t(brain_cancer_cnv.targets_T)))))

```

```{r}
# Wilcoxon test
# H0: The distribution of the effective targets (cnv values) is equal to the one of the cnv dataset
# H1: The distribution of the effective targets = cnv values are "" than the one of the cnv dataset
wilcox_c <- vector(mode = "list",length=ncol(brain_cancer_cnv.targets_T))
for(i in 1:ncol(brain_cancer_cnv.targets_T)) {         
  wilcox_c[[i]] <- wilcox.test(as.numeric(brain_cancer_cnv_T[,i]),as.numeric(brain_cancer_cnv.targets_T[,i]), alternative = "two.sided", paired = F)
}

wilcox_pvalue_c <- vector(length = ncol(brain_cancer_cnv.targets_T))
for(i in 1:ncol(brain_cancer_cnv.targets_T)) {
  wilcox_pvalue_c[i] <- wilcox_c[[i]]$p.value
}

length(which(wilcox_pvalue_c<=0.05))# 24 (less) /// 41 (two.sided)


```
##Adjustment of multiple testing problem 
```{r}
# Bonferroni correction
wilcox_pvalue_c_bonf <- p.adjust(wilcox_pvalue_c, method = "bonferroni", n = length(wilcox_pvalue_c)) 
b_correction_c <- which(wilcox_pvalue_c_bonf<=0.01)
p_values_c_b <- wilcox_pvalue_c_bonf[b_correction_c]
sig_targets_c <- colnames(brain_cancer_cnv.targets_T)[b_correction_c]

hist(p_values_c_b)
length(p_values_c_b)# 4 (less) /// 6 (two.sided)
  
# Benjamini-Hochberg correction
wilcox_pvalue_c_BH <- p.adjust(wilcox_pvalue_c, method = "BH", n = length(wilcox_pvalue_c))
bh_correction_c <- which(wilcox_pvalue_c_BH<=0.01)
```


















