---
title: 'Topic 1 Group 3: Drug viability screens for oncological and non-oncological
  treatments'
author: "Cedrik Neber, Lea Ahrens, Lennard Kleemann, Ilya Schneider, Xenia Quaas"
date: "19th July 2021"
output:
  pdf_document: default
  latex_engine: pdflatex
---

```{r packages, eval=TRUE, include=FALSE}
# packages
library(ggplot2)
library(dplyr)
library(factoextra)
```

```{r datasets, echo=FALSE}
load("../2021-topic-01-team-03/dataset/prism_datasets.rda")
load("../2021-topic-01-team-03/dataset/cellline_datasets.rda")
effective_in_all_doses=read.csv("../2021-topic-01-team-03/dataset/effective_in_all_doses.csv")
effective_in_all_doses=as.vector(effective_in_all_doses[,-1])
load("../2021-topic-01-team-03/dataset/ae_d7.RData")
load("../2021-topic-01-team-03/dataset/training.RData")
```

```{r cleanup, echo=FALSE}
# Cleanup
brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") 
names=brain_cancer_cl[,1]
brain_cancer=prism[names,]
for(i in 1:ncol(brain_cancer)){
  brain_cancer[is.na(brain_cancer[,i]), i] <- mean(brain_cancer[,i], na.rm = TRUE)
}
brain_cancer=brain_cancer[ , colSums(is.na(brain_cancer)) == 0]
brain_cancer_exp=prism.exp[names,]
```

Drug development is a time and investment consuming procedure. To tackle this problem, various new strategies have been explored. Remdesivir, originally developed to fight Ebola, has recently been employed to treat COVID-19 infected patients. The process of using an existing drug to combat an alternate disease is known as drug repurposing. 

Numerous advantages of this method include the reduced time and decreased likelihood to fail in research and development. Using the previously gathered data from a known drug additionally reduces the overall costs. In this context, computational approaches are gaining importance, as they allow large amounts of information from repurposing assays to be analyzed (Pushpakom *et al.*, 2019).

According to the Global Cancer Statistics 2020, brain cancer is a rare cancer type, accounting for 2,5% of all new cancer cases. Nevertheless, its mortality rate is comparatively high, which makes drug repurposing a promising approach to treat this disease.


# Structure of our project

This project uses seven datasets generated by the Broad Institute using the PRISM strategy. The aim is to investigate whether it is possible to predict the effectiveness of a drug in brain cancer treatment.

To explore this question, four milestones were examined:

- **How can we distinguish the most effective drugs?**

- **Are there any genetic markers that are specific for brain cancer subtypes?**

- **What are the targets of the effective drugs?**

- **What other factors contribute to drug and effectiveness prediction?**

# Data Clean up

As the datasets also contain information about other cancer types, irrelevant cell lines were discarded. Furthermore, the datasets were classified in accordance with the brain cancer subtypes. Cell lines with NA values were either removed or replaced with a mean value, depending on the dataset. The third milestone takes a closer look into brain cancer subtypes. To facilitate this step, additional data frames containing the individual subtypes were created.---


# Identification of effective drugs

The `prism` data frame gives information on treatments which were used in the original effectiveness screening. These treatments include 4,518 drugs with dosages ranging from 0.00061034 $\mu$M to 10 $\mu$M. Analysis of the data identified 8 standard dosages, which were then divided into separate data frames and stored in a list. 

Several dosages were identified, that did not correspond to the standard set of dosages. These outliers were assigned to the dosage with the least deviation, respectively. The idea of the assignment is to analyse the provided data via the drugs, as opposed to the assigned dosages. In this data set, the lowest values indicate the highest effectiveness. To identify the most effective drugs, the values of the `prism` data frame were analysed. Contrary to the threshold value 0.3 of Corsello *et al.* the threshold in this investigation was set to 0.2. This value was found to correspond to the median of the ```brain_cancer``` data frame (filtered brain cancer cell lines from the ```prism``` data frame), allowing the removal of over half of the provided drugs in this step. 

In order to reduce the number of drugs and to be able to make accurate predictions later on, only drugs were selected that are effective in all doses. First, the drugs whose mean effectiveness across all cell lines in the dose subset data frames was less than or equal to the established threshold were identified. Second, these drugs were compared to each other and only drugs present in all subsets were kept, allowing the identification of **51 drugs**. These drugs were saved in the ```effective_in_all_doses``` vector.
Subsequently, a single dosage was chosen for the further analysis, as the model aims to establish the output of drugs for suited for treatments, regardless of their concentration. Dosage 7 was selected due to its high variance in comparison to the other dosages, of 10 $\mu$M. 

```{r echo= FALSE,fig.dim= c(6,3), fig.align='left', fig.cap= 'Average subtype effectivness per drug'}

medulloblastoma=brain_cancer_cl$DepMap_ID[grep("Medulloblastoma", brain_cancer_cl$disease_subtype)]#we obtain the names of the cell lines that are classified as medulloblastoma
glioblastoma=brain_cancer_cl$DepMap_ID[grep("Glioblastoma", brain_cancer_cl$disease_subtype)]
glioma=brain_cancer_cl$DepMap_ID[grep("Glioma", brain_cancer_cl$disease_subtype)]
astrocytoma=brain_cancer_cl$DepMap_ID[grep("Astrocytoma", brain_cancer_cl$disease_subtype)]

glioma_ae_d7=ae_d7[glioma,]
glioblastoma_ae_d7=ae_d7[glioblastoma,]
astrocytoma_ae_d7=ae_d7[astrocytoma,]
medulloblastoma_ae_d7=ae_d7[medulloblastoma,]


## calculate the means for each drug in a subtype 
astrocytoma_ae_d7_mean = colMeans(astrocytoma_ae_d7)
glioma_ae_d7_mean = colMeans(glioma_ae_d7)
glioblastoma_ae_d7_mean = colMeans(glioblastoma_ae_d7)
medulloblastoma_ae_d7_mean = colMeans(medulloblastoma_ae_d7)

## create a combined data frame with all the means from each subtype 
means_eff <- data.frame(Astrocytoma = astrocytoma_ae_d7_mean,
                        Glioma = glioma_ae_d7_mean, 
                        Glioblastoma = glioblastoma_ae_d7_mean,  
                        Medulloblastoma = medulloblastoma_ae_d7_mean)
rownames(means_eff) = effective_in_all_doses

means_drugs <- data.frame(means_eff, 
                          Means = apply(means_eff, 1 , mean))

new_means_eff <- data.frame(Effectiveness = c(means_drugs$Astrocytoma, means_drugs$Glioma, means_drugs$Glioblastoma, means_drugs$Medulloblastoma), 
                            Drug_Mean = c(means_drugs$Means, means_drugs$Means, means_drugs$Means, means_drugs$Means), 
                            subtype = c(rep("Astrocytoma", nrow(means_drugs)),
                                      rep("Glioma", nrow(means_drugs)),
                                      rep("Glioblastoma", nrow(means_drugs)),
                                      rep("Medulloblastoma", nrow(means_drugs))), 
                            Drug = c(effective_in_all_doses, effective_in_all_doses, effective_in_all_doses, effective_in_all_doses))


## rearrange the Values in descending order 
first = new_means_eff$Drug_Mean
second = sort(new_means_eff$Drug_Mean, decreasing = TRUE)
              


arrangement <- data.frame(Drug_Ranks = match(first, second), 
                                Effectiveness = new_means_eff$Effectiveness, 
                                Subtype = new_means_eff$subtype,
                                Drug = new_means_eff$Drug, 
                                Drug_Mean = new_means_eff$Drug_Mean)
res <- 1 

for(i in c(1:51)){zv <- rep(i,4);res <- c(res, zv)}
vec <- res[2:205]
pre_arrangement <- arrange(arrangement, Drug_Ranks)
pre_arrangement <- data.frame(pre_arrangement, 
                               Numbers = vec)

## Plot the new, rearranged subtypes 
ggplot(pre_arrangement, aes(x = Numbers , y = Effectiveness, color = Subtype)) + 
                          xlab("Drugs") + 
                          ylab("Effectiveness") + 
  
  geom_point(aes(shape=Subtype))
```

Figure 1 represents the average effectiveness of each drug for each subtype. The means of the identified drugs have effectiveness scores ranging from 0.2 to  approximately -8.2. In the graph shown, the drugs were ranked to show the trend int heir effectiveness. Overall, there is general tendency for medulloblastoma cell lines to be more effected by the drug treatment. For 31 out of the 51 drugs, medulloblastoma cell lines shown the highest effectiveness scores. However, these could be attributed to the fact that medulloblastoma, with only two cell lines, as opposed to the 20 glioblastoma cell lines. 


# Identification of genetic markers

For this milestone, a further look at brain cancer subtypes was taken. This information could be relevant for the final regression model. The column ```disease_subtype``` of the ```prism.cl``` dataset gives information about the subtype, whereby a distinction is made between astrocytoma, medulloblastoma, glioblastoma and glioma. 

```{r echo=FALSE, fig.dim= c(6,3), fi.align= 'left', fig.cap= 'Gene expression clustering'}
#PCA before applying the kmeans 

exp <- brain_cancer_exp
exp_pca <- prcomp(exp)


## kmeans for the clustering.frame 

set.seed(126)

PC <- as.data.frame(exp_pca$x)
PC1 <- PC$PC1
PC2 <- PC$PC2

## Proper kmeans after PCA 

km_c <- kmeans(exp_pca$x, 4, iter.max = 10000, nstart = 1)

## create a new data frame for the ggplot visualisation 
gg_pca <- data.frame( PC1 = PC1,
                      PC2 = PC2, 
                      Cell_line = rownames(exp), 
                      Subtype = brain_cancer_cl$disease_subtype, 
                      Cluster = as.factor(km_c$cluster))

ggplot(gg_pca, aes( x = PC1, y = PC2, shape = Cluster, colour = Subtype)) +
  xlab("PC1") + 
  ylab("PC2")+ 
  geom_point(size = 4)

```

Figure 2 shows a clustering of the two principle components (PC) with the highest variance of the cell lines by subtype. The data used is the expression data from the brain cancer subtypes. In the first step a principle component analysis (PCA) was conducted to reduce the dimensionality of the data frame before using the kmeans method to cluster the obtained PC's. This method divides observations into clusters by assigning them to the cluster with the smallest euclidean distance to a cluster centroid. The number of expected clusters was predefined as 4. This was chosen as the graph aims to analyse the relationship of the gene expression of the cell lines and the subtypes they belong to. 

As can be seen, there is no link between the clusters and the subtypes, since there is no pattern between the symbols and the colours in the plot. This makes the continuous analysis using the subtypes obsolete. Furthermore, the clustering in itself does not seem to be conclusive (see cluster 1). Even an alternate number of itertions in the kmeans did not change that. 

## Shapiro Wilks test

To find out which genes are expressed significantly differently in brain cancer cell lines compared to other cancers, a statistical test can be applied. For this purpose, it must first be examined whether the available data is normally distributed.
A statistical test, the Shapiro Wilks test, can be used for this evaluation. In this case, the H0 hypothesis states that the distribution is a normal distribution. If the p-value is smaller than the significance level, the H0 hypothesis is rejected and it is assumed that the data is not normally distributed. 
The test was applied to the 19177 genes from the ```prism.exp``` dataset to determine the expression across the 34 cell lines. A significance level of 0.05 was set. As a result, 9799 p-values were less than or equal to this value, thus approximately half of the genes were normally distributed. This is the reason a non-parametric test was conducted.  
\newpage

## Wilcoxon Rank Sum test

An unpaired Wilcoxon Rank Sum test was used to identify differential gene expression. In this test two different datasets were compared. The H0 hypothesis is that expression values of both samples come from the same distribution.  The first dataset was ```non_bc_exp```, which has all the cell lines from the ```prism.exp``` dataset except from the brain cancer disease type. There are 447 cell lines in total from lung, skin, pancreatic and ovarian cancers. The second dataset is the ```brain_cancer_exp```, which reflects the level of gene expression of the 34 brain cancer cell lines.  Although these data sets are not identical in size, this is not a problem, as the test is based on ranking the individual values. Some genes have an expression value of 0. To avoid following complications (for example log2-fold-change calculation) the lowest expression value for each dataset, so called ```pseudo_count```, was added. The pseudo values were equal for both datasets.


```{r echo=FALSE}
# Creating dataframe with all non brain cancer cell lines
non_bc_cl = prism.cl[!prism.cl$disease == "Brain Cancer",]
non_bc_names=non_bc_cl[,1] 
non_bc_exp=prism.exp[non_bc_names,]
brain_cancer_exp_w=brain_cancer_exp
non_bc_exp=non_bc_exp[rowSums(is.na(non_bc_exp)) == 0,]



#Z-transformation
calculate_pseudo_v<-function(x,y){
   min(x[y,][which(x[y,]>0)])
}
#function to determine the pseudovalue for both datasets. It takes a dataframe and looks for a minimum value for each row (in this case cell line). Then it chooses only the minimum values that are bigger than 0. Out of these positive mins, it chooses the smallest one, which is our pseudo value

pseudo_value1<-calculate_pseudo_v(non_bc_exp,nrow(non_bc_exp))
pseudo_value2<-calculate_pseudo_v(brain_cancer_exp_w, nrow(brain_cancer_exp_w))
#Now we have 2 pseudo values from 2 datasets, we wil compare them in the following loop to determine the ultimate pseudo value to apply on both datasets

if(pseudo_value1==pseudo_value2){
  pseudo_value<-pseudo_value1
} else if(pseudo_value1>pseudo_value2) {
  pseudo_value<-pseudo_value2
} else {
  pseudo_value<-pseudo_value1
}


non_bc_exp=non_bc_exp+pseudo_value
brain_cancer_exp_w=brain_cancer_exp_w+pseudo_value

non_bc_exp=as.data.frame(t(scale(as.matrix(t(non_bc_exp)))))
brain_cancer_exp_w=as.data.frame(t(scale(as.matrix(t(brain_cancer_exp_w)))))

# Removal of lines with NA from new dataframe
brain_cancer_exp_w=as.data.frame(t(na.omit(t(brain_cancer_exp_w))))
non_bc_exp=non_bc_exp[,colnames(brain_cancer_exp_w)]
```

```{r eval = TRUE, include=FALSE}

# Performance of Wilcoxon test

wilcox_genes <- vector(mode = "list",length=ncol(brain_cancer_exp_w))
for(i in 1:ncol(brain_cancer_exp_w)) {         
  wilcox_genes[[i]] <- wilcox.test(as.numeric(non_bc_exp[,i]),
                                   as.numeric(brain_cancer_exp_w[,i]), 
                                   alternative = "two.sided", paired = F)
}
```

In this non-parametric test the H0 hypothesis is tested simultaneously for all genes. This leads to a multiplicity problem, in which the probability of increasing false positive test results is increased. To adjust the number of significant p-values occurring for this problem, we consider two types of corrections. The Benjamini-Hochberg and the Bonferroni corrections. 

The number of genes that are significantly differentially expressed differs drastically. The Benjamini-Hochberg correction results in 8948 genes, meanwhile the Bonferroni correction in **1859 genes**. The Bonferroni correction was chosen, since it is more conservative and controls the type I error. Therefore the number of false positives (keeping genes that are not significant) is avoided. 

```{r echo=FALSE, fig.dim= c(5,4), fig.cap= 'Volcano plot of significant differential expressed genes'}

#P values of Wilcoxon test genes
wilcox_pvalue_genes <- vector(length = ncol(brain_cancer_exp_w))
for(i in 1:ncol(brain_cancer_exp_w)) {
  wilcox_pvalue_genes[i] <- wilcox_genes[[i]]$p.value
}

#Bonferroni correction
wilcox_pvalue_bonf = p.adjust(wilcox_pvalue_genes, method = "bonferroni", n = length(wilcox_pvalue_genes)) #I first was confused from p-value=1 but is possible when there is absolutely no evidence to reject H0
b_correction=which(wilcox_pvalue_bonf<=0.01)
p_values_b=wilcox_pvalue_bonf[b_correction]

non_bc_exp=prism.exp[non_bc_names,]

volcano_bc=brain_cancer_exp[,b_correction]
volcano_non_bc=non_bc_exp[,b_correction]

df<-data.frame(matrix(nrow = length(b_correction),ncol = 2))
rownames(df)=colnames(volcano_bc)
colnames(df)=c("-log10 p-value", "log2 foldchange")
df[,1]=-log10(p_values_b)

bc_exp_mean=apply(volcano_bc,2, mean)
non_bc_exp_mean=apply(volcano_non_bc,2,mean, na.rm=T)

df[,2]=log2((non_bc_exp_mean/bc_exp_mean))
logfc=df[,2]

highlight_df <- df %>% filter(logfc>=1|logfc<=-1)
bp<-sort(highlight_df[,1],decreasing = T)[1:20]
best_p_values<-data.frame(matrix(nrow =length(bp),ncol = 2))
best_p_values<-data.frame(matrix(nrow =length(bp),ncol = 2))
for (y in 1:length(bp)) {
  best_p_values[y,]<-highlight_df[which(highlight_df[,1]==bp[y]),]
  rownames(best_p_values)[y]=rownames(highlight_df[which(highlight_df[,1]==bp[y]),])
}


ggplot(df, aes(x=df[,2], y=df[,1]),ylim = range(0,1))+
  xlab("Log2-Fold-Change") + 
  ylab("-log10(p-value)") + 
  geom_point() +
  geom_point(data=highlight_df, aes(x=highlight_df[,2], y=highlight_df[,1]), color='blue') +
  geom_point(data=best_p_values, aes(x=best_p_values[,2], y=best_p_values[,1]), color='red')
top_genes=rownames(best_p_values)
```

To visualizes differentially expressed genes a volcano plot can be used, as shown on the Figure 3. Each gene expression difference can be represented by the log2-fold-change on the x-axes. The y-axes displays the -log10 of each p-value. First, the genes were filtered according to the log2-fold-change. Only the genes, which value was less or equal to -1 and greater or equal to 1 were kept. These genes are shown in blue on the plot. Following that, the top 20 genes with the biggest -log10 p-value were extracted and highlighted in red.
\newpage

## Linear regression model with universal drug

From these top 20 genes, the ones with the most influence on drug effectiveness were selected. An effective drug from dosage 7 that has an overall consistent effect on the cell lines needs to be determined.  Therefore, a "universal drug" was filtered out by dividing the range of effectiveness scores into equally sized intervals. For each drug an interval with the highest amount of responding cell lines was selected. The determined universal drug is the one with the biggest count (Broad ID "BRD-K79145628-001-05-5"). 

The same method was applied to the non brain cancer cell lines and after that two universal drugs were compared. They turned out to be different, which indicated that the brain cancer universal drug is specific. 

This drug was finally used as a reference variable for the linear regression model, in which drug effectiveness is predicted by the expression of the top 20 genes. In this case, each gene is considered as a single variable. By setting up a multiple regression, it is important to note that no correlated variables should be used. Among the top 20 genes, a high correlation between the genes "FAM83H" and "IQANK1" was found. The gene with a more significant p-value in the Wilcoxon test was kept. 

The application of the linear regression model involves feature selection. It consists of a reduction in the number of variables, so that only those that contribute the most to the prediction are included in the model. The p-value of the F-test was used for the evaluation, since it compares the linear model with the null model. In the F-test the H0 hypotheses is that the tested model does not perform better than the null model, therefore a small p-value leads to the rejection of the H0 hypotheses. 

```{r eval=TRUE, include=FALSE}

# Computing the global minimum and maximum
min_eff=min(apply(ae_d7,2,min))
max_eff=max(apply(ae_d7,2,max))

# Creating a sequence for the interval
interval=seq(from=min_eff, to=max_eff, by=((abs(min_eff)+max_eff))/20)

# Put the data frame row wise in the intervals
eff_intervals <- vector(mode = "list",length=ncol(ae_d7)) 
for(i in 1:ncol(ae_d7)) {  
  eff_intervals[[i]]=table(cut(ae_d7[,i],breaks=interval,right=TRUE,include.lowest=TRUE))
  }

# Extracting the maximum value and so the universal drug
vec_max_eff=unlist(lapply(eff_intervals,max))
uni_drug=which(vec_max_eff==max(vec_max_eff))
colnames(ae_d7)[uni_drug]
```

```{r eval=TRUE, include=FALSE}
# Creating a data frame with top 20 genes and universal drug
highlight_df <- df %>% filter(logfc>=1|logfc<=-1)
bp<-sort(highlight_df[,1],decreasing = T)[1:20]
best_p_values<-data.frame(matrix(nrow =length(bp),ncol = 2))
for (y in 1:length(bp)) {
  best_p_values[y,]<-highlight_df[which(highlight_df[,1]==bp[y]),]
  rownames(best_p_values)[y]=rownames(highlight_df[which(highlight_df[,1]==bp[y]),])
}
top_genes=rownames(best_p_values)
genes_regression=cbind(brain_cancer_exp[,top_genes],ae_d7[,uni_drug])
colnames(genes_regression)[ncol(genes_regression)]="Drug"

```

```{r First linear regression model, results = 'hide'}

# linear regression model
initial_regression_1<-summary(lm(Drug~.,genes_regression)) # 1. (model A)

# feature selection
repeat{
  end_regression_1<-initial_regression_1
  
  pvx<-pf(end_regression_1$fstatistic[1],end_regression_1$fstatistic[2],
          end_regression_1$fstatistic[3],lower.tail=FALSE)
  
coeffs=as.data.frame(end_regression_1$coefficients)
coeffs=coeffs[-c(1),]

coeffs=coeffs[order(coeffs$`Pr(>|t|)`),] # 2.
coeffs=coeffs[-c(nrow(coeffs)),]
genes_regression=cbind(brain_cancer_exp[,rownames(coeffs)],ae_d7[,uni_drug])
colnames(genes_regression)[ncol(genes_regression)]="Drug"

initial_regression_1<-summary(lm(Drug~.,genes_regression)) # 3.(model B)
pvy<-pf(initial_regression_1$fstatistic[1],initial_regression_1$fstatistic[2],
        initial_regression_1$fstatistic[3],lower.tail=FALSE)

if(pvy>=pvx){
  break # 4.
  }
}

end_regression_1 # 5.

final_genes=rownames(as.data.frame(end_regression_1$coefficients))
final_genes=final_genes[-1] # 6.
```
\newpage

The following strategy was applied:

1. All 20 genes were put in the model (model A).
2. Genes were ordered according to their own p-values from the highest to the lowest.
3. The lowest one was removed and the new linear model was performed (model B).
4. If p-value of model A is bigger than p-value of model B, the procedure is repeated starting from the second step.
5. This loop stops, the moment p-value of model A is smaller than p-value of model B.
6. The model has reached its optimal point.

**7 final genes** arise from this model with a p-value of 1,353%. All 447 cell lines underwent the same strategy with the linear regression. As a result a drastically lower r-squared value with only 3 genes was determined to be the optimal model in this case. This indicates the specificity of these 20 genes in the brain cancer cell lines.

To further check this result, the full model is compared with the reduced model using the ```anova``` function. It compares the variances of the residuals considering the introduced degrees of freedom. The H0 hypothesis is that the two models are equivalent. A p-value of **0.9446** is obtained, which suggests that the reduced model performs significantly better in comparison with the full one. 

# Identification of drug targets

The 51 effective drugs which were identified in question one were used to determine their specific targets. The ```prism.treat``` delivered this information. There were drugs which attack more than one target or had no target available. Some targets occur for different drugs. In total, **124 targets** for all effective drugs remained. 

To explore these targets, three other data frames where used. The  ``` brain_cancer_achilles ``` gives information on gene knockdown scores what indicates the importance for cell survival. The ``` brain_cancer_cnv ``` contains gene copy number values and therefore reflects genetic alterations. Lastly, the ``` brain_cancer_exp ``` dataset was used, which was already used for the identification of genetic markers. Not for all targets were all values available, therefore "ATP5A1", "MMP12" and "MMP23A" were not utilized for further analysis.

## Analysis of the targets
To find out whether the distribution of the target values from one of the datasets had the same distribution as one of the complete datasets, a QQ-plot was applied. This is a graphical analysis which compares two probability distributions by their quantiles. 


```{r  echo=FALSE, fig.dim= c(5,4),  fig.cap= c("Comparison of the distribution of proliveration values of the target knockout of the effective targets with all targets","Comparison of the distribution of expression values of the targets of the effective targets with all targets", "Comparison of the distribution of copie number values of the target knockout of the effective targets with all targets")}

brain_cancer_treat=prism.treat[colnames(brain_cancer),] #selecting only the drugs we previously filtered out from brain_cancer
brain_cancer_treat=brain_cancer_treat[!duplicated(brain_cancer_treat$broad_id),]
rownames(brain_cancer_treat)<-brain_cancer_treat[,1]
effective_bc_treat=brain_cancer_treat[effective_in_all_doses,-c(1,2,3,4,10)]

brain_cancer.effective_targets <- effective_bc_treat$target

# which(is.na(effective_bc_treat$target))


brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") #identify the brain cancer cell lines' names

names=brain_cancer_cl[,1] #saving the names as names

#select wanted cell lines from all datasets
brain_cancer=prism[names,]
brain_cancer_achilles=prism.achilles[names,]

brain_cancer_cnv=prism.cnv[names,]
brain_cancer_exp=prism.exp[names,]


brain_cancer.effective_targets <- as.character(brain_cancer.effective_targets) # convert to character

x <- 1
brain_cancer.effective_targets.solo <- list() #create list

for(i in brain_cancer.effective_targets) #split multiple targets
  {
  
 brain_cancer.effective_targets.solo[[x]] <- strsplit(i, split = ", ")
 x <- x+1
 }
brain_cancer.effective_targets.solo <- unlist(brain_cancer.effective_targets.solo) #list to vector

brain_cancer.effective_targets.solo <- unique(brain_cancer.effective_targets.solo) # remove multiple targets

brain_cancer.effective_targets.solo <- brain_cancer.effective_targets.solo[-1] # remove NA


# achilles targets
brain_cancer_achilles.targets <- brain_cancer_achilles[,which(colnames(brain_cancer_achilles) %in% brain_cancer.effective_targets.solo)] # selecting target genes in achilles
#brain_cancer.effective_targets.solo[-which(brain_cancer.effective_targets.solo %in% colnames(brain_cancer_achilles))] # missing targets
qqplot(unlist(brain_cancer_achilles), unlist(brain_cancer_achilles.targets), xlab = "Complete prism.achilles", ylab = "prism.achilles values of the effective targets")
abline(a = 0, b = 1)

# exp targets
brain_cancer_exp.targets <- brain_cancer_exp[,which(colnames(brain_cancer_exp) %in% brain_cancer.effective_targets.solo)] # selecting target genes in exp
#brain_cancer.effective_targets.solo[-which(brain_cancer.effective_targets.solo %in% colnames(brain_cancer_exp))] # missing targets
qqplot(unlist(brain_cancer_exp), unlist(brain_cancer_exp.targets), xlab = "Complete prism.exp", ylab = "prism.exp values of the effective targets")
abline(a = 0, b = 1)

# cnv targets
brain_cancer_cnv.targets <- brain_cancer_cnv[,which(colnames(brain_cancer_cnv) %in% brain_cancer.effective_targets.solo)] # selecting target genes in cnvVie
#brain_cancer.effective_targets.solo[-which(brain_cancer.effective_targets.solo %in% colnames(brain_cancer_cnv))] # missing targets
qqplot(unlist(brain_cancer_cnv), unlist(brain_cancer_cnv.targets), xlim = c(0,2.5), ylim = c(0,2.5), xlab = "Complete prism.cnv", ylab = "prism.cnv values of the effective targets")
abline(a = 0, b = 1)

```

```{r echo=FALSE, fig.dim= c(5,4), fig.cap= c("Comparison of the distribution of proliveration values of the target knockout of the effective targets with all targets")}

qqplot(unlist(brain_cancer_achilles), unlist(brain_cancer_achilles.targets), xlab = "Complete prism.achilles", ylab = "prism.achilles values of the effective targets")
abline(a = 0, b = 1)

```

```{r  echo=FALSE, fig.dim= c(5,4), fig.cap= c("Comparison of the distribution of expression values of the targets of the effective targets with all targets")}

```

```{r  echo=FALSE, fig.dim= c(5,4), fig.cap= c("Comparison of the distribution of copie number values of the target knockout of the effective targets with all targets")}

```


# Plots nebeneinander mit: fig.show="hold", out.width="30%"

The visualizations made clear, that the distribution of the effective targets from .exp and .achilles are not representative for the complete dataset. A similar shape of both distribution can just be accepted if the QQ-plot is a straight line. Therefore, these two parameters seem to have a correlation to the effectiveness.

To further analyze this correlation another Wilcoxon Rank Sum test was applied. For the .achilles a one-sided test was made and resulted in 59 targets out of 121, which derive from the distribution of the complete dataset.  After applying the Bonferroni correction only **44 targets** left. For the .exp and the .cnv data a two-sided Wilcoxon test was applied. The corrected values are **41 targets** for .exp and **6 targets** for .cnv, which did not confirm the H0-hypothesis. In conclusion, the effectiveness correlates in some way with the .achilles and the .exp data. However, there are still other parameters, which contribute to the effectiveness for specific target.

# Linear regression model for effectivness prediction

The fist idea was to predict the effectiveness based on a chosen cell line and a drug. But this idea was discarded because the three datasets ```prism.achilles```, ```prsim.exp ``` and ``` prism.cnv``` depend only on the targets and not the drugs. 
Therefore, the decision was made to predict the effectiveness based on a chosen cell line and one target. Hence, the value for one target of each of the three datasets is specific and depends on a single cell line. The ```prism``` values, which give information about the effectiveness, were averaged, because more than one drug can have the same target. The focus was only set on the highest drug dosage.

Lm

```{r}
training <- na.omit(training) # delete 66 NA rows

pairs(~prism+IQANK1+RAB11FIP1+FOXG1+LPIN3+ERVMER34.1+PTN+VRK2+achilles+exp+cnv, data = training, pch = 20, cex = 0.5) # heatmap?

devide <- sample(1:nrow(training),nrow(training)/2)

dat.train <- training[devide,]
dat.test <- training[-devide,]

lm1 <- lm(prism~IQANK1+RAB11FIP1+FOXG1+LPIN3+ERVMER34.1+PTN+VRK2+achilles+exp+cnv+prism, data = dat.train)
summary(lm1)


pca <- prcomp(dat.train[,-c(11)])
summary(pca)
for(i in 1:(ncol(dat.train)-1)){
par(las=2);barplot(pca$rotation[,i],horiz = TRUE, main = paste("PC",i))
}


lm.pca <- lm(dat.train$prism~pca$x)
summary(lm.pca)





pred <- predict(l2, newdata = dat.test)
plot(dat.test$prism,pred)
abline(a=0,b=1, col="red")


n <- nrow(dat.train)
rmse.test <- sqrt(1/n * sum(l2$residuals^2))
rmse.test


n <- nrow(dat.test)
residuals <- dat.test$prism - pred
rmse.test <- sqrt(1/n * sum(residuals^2))
rmse.test


l3 <- lm(prism~IQANK1+RAB11FIP1+FOXG1+LPIN3+ERVMER34.1+PTN+VRK2+achilles+exp+cnv+prism, data = training1)
summary(l3)






# Reduce regression model

initial_regression_1 <- summary(l.pca3)


repeat{
  end_regression_1<-initial_regression_1
  pvx<-pf(end_regression_1$fstatistic[1],end_regression_1$fstatistic[2],
          end_regression_1$fstatistic[3],lower.tail=FALSE)
  
coeffs=as.data.frame(end_regression_1$coefficients)
coeffs=coeffs[-c(1),]
coeffs=coeffs[order(coeffs$`Pr(>|t|)`),] 
coeffs=coeffs[-c(nrow(coeffs)),]

pca.left <- sapply(row.names(coeffs),function(x){unlist(strsplit(x, split = "x"))[2]})
pca3.leftx <- pca3$x[,which(colnames(pca3$x) %in% pca.left)] 

initial_regression_1<-summary(lm(training1$prism~pca3.leftx)) 
pvy<-pf(initial_regression_1$fstatistic[1],initial_regression_1$fstatistic[2],
        initial_regression_1$fstatistic[3],lower.tail=FALSE)
if(pvy>=pvx){
  break 
}
}

initial_regression_1


anova(lm(training1$prism~pca3.leftx),l.pca3)
```


To reduce the amount of input values effectively with constant p-value and constant proportion of the explained variance, a PCA was applied. The calculated PC values were used to train the model. 

PCA

It is visible that the model only describes little proportion of the variance. However, the regression model is still better than the null model. A reason for that could be, that the data offers known information and drug repurposing is mostly based on unknown effective mechanisms. Maybe a more complex method of machine learning could find more correlation between our given parameters. Because such a model only bases on a specific amount of parameters it only can support real experiments.


# Discussion

The universal drug is called nimorazole and is an anti-infective drug which acts as bacterial DNA inhibitor. Moreover, it can be used as radiosensitizer to improve radiotherapy in head and neck squamous cell carcinoma. A radioresistance results in many cancer types because of a low oxygen level. Nimorazole enhances cell death by binding to damaged DNA and preventing repair mechanisms. 

The brain cancer subtype glioma can be divided in astrocytoma and glioblastoma. We suspect that the description glioma is occuring whenever there is a classification missing. As these three subtypes belong to the same class of cancers and the clustering did not lead to another result, we decided on not take the subtypes into account. 

A great promise of combining precision medicine with drug repurposing is to avoid the use of drugs which are ineffective in patients and even could lead to side effects. Moreover, drug repurposing leads to a cost reduction. According to Workman et al. a cancer drug in the US costs $100 ,000 anually in 2012 and the tendency is increasing. 

Involving drug targets for the prediction of the effectiveness of brain cancer treatment is problematically. As Maughan (2017) pointed out, cancer is caused by a complex interplay of multiple genetic factors. The high expectation of finding specific targets for cancer therapies is an imagination caused by a few successful targeted drugs like trastuzumab for breast cancer. Nevertheless, a problem of tumors is the heterogeneity of cells. A resistance against a drug can results when the it adapts to the drug. 


# Literature

Chen, R., Smith-Cohn, M., Cohen, A.L., and Colman, H. (2017). Glioma Subclassifications and Their Clinical Significance. Neurotherapeutics 14, 284-297.

Corsello, S. M., Nagari, R. T., Spangler, R. D., Rossen, J., Kocak, M., Bryan, J. G., … Golub, T. R. (2020). Discovering the anticancer potential of non-oncology drugs by systematic viability profiling. Nature Cancer, 1(2), 235–248.

DeWeerdt, S. (2018). The genomics of brain cancer. Nature, 561(7724), 54-55.

Liang, C., Tian, L., Liu, Y., Hui, N., Qiao, G., Li, H., Shi, Z., Tang, Y., Zhang, D., Xie, X., Zhao, X., A promising antiviral candidate drug for the COVID-19 pandemic: A mini-review of remdesivir. (2020). Eur J Med Chem., 201, 112527. 

Fang, Z., Du, R., and Cui, X. (2012). Uniform approximation is more appropriate for Wilcoxon Rank-Sum Test in gene set analysis. PLoS One 7, e31505.

Pushpakom, S., Iorio, F., Eyers, P. A., Escott, K. J., Hopper, S., Wells, A., … Pirmohamed, M. (2019). Drug repurposing: progress, challenges and recommendations. Nature Reviews Drug Discovery, 18(1), 41–58.

Sung, H., Ferlay, J., Siegel, R.L., Laversanne, M., Soerjomataram, I., Jemal, A., and Bray, F. (2021). Global Cancer Statistics 2020: GLOBOCAN Estimates of Incidence and Mortality Worldwide for 36 Cancers in 185 Countries. CA Cancer J Clin 71, 209-249.

Tan, Y.D., and Xu, H. (2014). A general method for accurate estimation of false discovery rates in identification of differentially expressed genes. Bioinformatics 30, 2018-2025.
