---
title: 'Topic 1 Group 3: Drug viability screens for oncological and non-oncological
  treatments'
author: "Cedrik Neber, Lea Ahrens, Lennard Kleemann, Ilya Schneider, Xenia Quaas"
date: "19th July 2021"
output:
  pdf_document: default
  html_document: default
---

```{r packages, eval=TRUE, include=FALSE}
# packages
library(ggplot2)
library(dplyr)
library(factoextra)
```

```{r datasets, echo=FALSE}
load("../2021-topic-01-team-03/dataset/prism_datasets.rda")
load("../2021-topic-01-team-03/dataset/cellline_datasets.rda")
#load("../2021-topic-01-team-03/dataset/effective_in_all_dosis.rda")
```

```{r cleanup, echo=FALSE}
# Cleanup
brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") 
names=brain_cancer_cl[,1]
brain_cancer_exp=prism.exp[names,]
```

Drug development is a time and investment consuming procedure. To tackle this problem various new strategy have been explored. Remdesivir, originally developed to fight Ebola, has recently been employed to treat COVID-19 infected patients. The process of using an existing drug to combat an alternate disease is known as drug repurposing. 

Numerous advantages of this method include the reduced time and decreased likelihood to fail in research and development. Using the previously gathered data from a known drug additionally reduces the overall costs. In this context, computational approaches are gaining importance, as they allow large amounts of information to be analyzed (Pushpakom *et al.*, 2019).

According to the Global Cancer Statistics 2020, brain cancer is a rarer cancer type, accounting for 2,5% of all new cancer cases. Nevertheless, its mortality rate is comparatively high, which makes drug repurposing a promising approach to treat this disease.


# Structure of our project

This project uses seven datasets generated by the Broad Institute using the PRISM strategy. The aim is to investigate whether it is possible to predict the effectiveness of a drug in brain cancer treatment.

To explore this question, four milestones were examined:

**1) How can we distinguish the most effective drugs?**

**2) What are the targets of the effective drugs?**

**3) Are there any genetic markers that are specific for brain cancer subtypes?**

**4) What other factors contribute to drug and effectiveness prediction?**

# Data Clean up

As the datasets also contain information about other cancer types, irrelevant cell lines were discarded. Furthermore, the datasets were classified in accordance with the brain cancer subtypes. Cell lines with NA values were either removed or replaced with a mean value, depending on the dataset. The third milestone takes a closer look into brain cancer subtypes. To facilitate this step, additional dataframes with regard to the subtypes were created.


# Identification of effective drugs

The `prism` data frame gives information on treatments which were used in the original effectiveness screening. These treatments include 4,518 drugs with dosages ranging from 0.00061034 $\mu$M to 10 $\mu$M. Analysis of the data identified 8 standard dosages, which were then divided into separate data frames and stored in a list. 

Several dosages were identified, that did not correspond to the standard set of dosages. These outliers were assigned to the dosage with the least deviation, respectively. The idea of the assignment is to analyse the provided data via the drugs, as opposed to the assigned dosages. In this data set, the lowest values indicate the highest effectiveness. To identify the most effective drugs, the values of the `prism` data frame were analysed. Contrary to the threshold value 0.3 of Corsello *et al.* the threshold in this investigation was set to 0.2. This value was found to correspond to the median of the ```brain_cancer``` data frame (filtered brain cancer cell lines from the ```prism``` data frame), allowing the discard of over half of the provided drugs in this step. 

In order to reduce the number of drugs and to be able to make later predictions, only drugs were selected that are effective in all doses. First, the drugs whose mean effectiveness across all cell lines in the dose subset data frames was less than or equal to the established threshold were identified. Second, these drugs were compared to each other and only drugs present in all subsets were kept, allowing the identification of **51 drugs**.
Subsequently, a single dosage was chosen for the further analysis, as the model aims to establish the output of drugs for suited for treatments, regardless of their concentration. Dosage 7 was selected due to its high variance, in comparison to the other dosages, of 10 $\mu$M. 

```{r eval=FALSE, echo=FALSE, fig.dim= c(5,4), fig.cap= 'Figure 1: Average subtype effectivness per drug'}

ae_d7 <-data.frame(matrix(nrow = 34,ncol = length(effective_in_all_doses)),row.names = rownames(brain_cancer))
colnames(ae_d7)=effective_in_all_doses

for (i in 1:length(effective_in_all_doses)){
  ae_d7[,i]=effective_d7[,grep(effective_in_all_doses[i], colnames(effective_d7))]
}

glioma_ae_d7=ae_d7[glioma,]
glioblastoma_ae_d7=ae_d7[glioblastoma,]
astrocytoma_ae_d7=ae_d7[astrocytoma,]
medulloblastoma_ae_d7=ae_d7[medulloblastoma,]


## calculate the means for each drug in a subtype 
astrocytoma_ae_d7_mean = colMeans(astrocytoma_ae_d7)
glioma_ae_d7_mean = colMeans(glioma_ae_d7)
glioblastoma_ae_d7_mean = colMeans(glioblastoma_ae_d7)
medulloblastoma_ae_d7_mean = colMeans(medulloblastoma_ae_d7)

## create a combined data frame with all the means from each subtype 
means_eff <- data.frame(Astrocytoma = astrocytoma_ae_d7_mean,
                        Glioma = glioma_ae_d7_mean, 
                        Glioblastoma = glioblastoma_ae_d7_mean,  
                        Medulloblastoma = medulloblastoma_ae_d7_mean)
rownames(means_eff) = effective_in_all_doses

means_drugs <- data.frame(means_eff, 
                          Means = apply(means_eff, 1 , mean))

new_means_eff <- data.frame(Efficiency = c(means_drugs$Astrocytoma, means_drugs$Glioma, means_drugs$Glioblastoma, means_drugs$Medulloblastoma), 
                            Drug_Mean = c(means_drugs$Means, means_drugs$Means, means_drugs$Means, means_drugs$Means), 
                            subtype = c(rep("Astrocytoma", nrow(means_drugs)),
                                      rep("Glioma", nrow(means_drugs)),
                                      rep("Glioblastoma", nrow(means_drugs)),
                                      rep("Medulloblastoma", nrow(means_drugs))), 
                            Drug = c(effective_in_all_doses, effective_in_all_doses, effective_in_all_doses, effective_in_all_doses))


## rearrange the Values in descending order 
first = new_means_eff$Drug_Mean
second = sort(new_means_eff$Drug_Mean, decreasing = TRUE)
              


arrangement <- data.frame(Drug_Ranks = match(first, second), 
                                Efficiency = new_means_eff$Efficiency, 
                                Subtype = new_means_eff$subtype,
                                Drug = new_means_eff$Drug, 
                                Drug_Mean = new_means_eff$Drug_Mean)
res <- 1 

for(i in c(1:51)){zv <- rep(i,4);res <- c(res, zv)}
vec <- res[2:205]
pre_arrangement <- arrange(arrangement, Drug_Ranks)
pre_arrangement <- data.frame(pre_arrangement, 
                               Numbers = vec)

## Plot the new, rearranged subtypes 
ggplot(pre_arrangement, aes(x = Numbers , y = Efficiency, color = Subtype)) + 
                         ggtitle("Means of subtypes") +
                          xlab("Drugs") + 
                          ylab("Efficiency") + 
  
  geom_point(aes(shape=Subtype))
```


**Figure 1** represents the average effectiveness of each drug for each subtype. The means of the identified drugs have effectiveness scores ranging from 0.2 to  approximately -10. There is a linear relationship between means' effectiveness and drugs. There is general tendency for medulloblastoma cell lines to be more effected by the drug treatment.However, these could be attributed to the fact that medulloblastoma, with only two cell lines, contributes the least amount of data.

# Identification of drug targets

Lea&Lennard 

+++QQ Plot+++

# Identification of genetic markers

For this milestone, a further look at brain cancer subtypes was taken. This information could be relevant for the final regression model. The column ```disease_subtype``` of the ```prism.cl``` dataset gives information about the subtype whereby a distinction is made between astrocytoma, medulloblastoma, glioblastoma and glioma. 

```{r eval=FALSE, echo=FALSE, fig.dim= c(5,4), fig.cap= 'Gene expression clustering'}
#PCA before applying the kmeans 

exp <- brain_cancer_exp
exp_pca <- prcomp(exp)
 
# plot the PC's an examine their variance 
fviz_eig(exp_pca)


## kmeans for the clustering.frame 

set.seed(126)

PC <- as.data.frame(exp_pca$x)
PC1 <- PC$PC1
PC2 <- PC$PC2

## Proper kmeans after PCA 

km_c <- kmeans(exp_pca$x, 4, iter.max = 10000, nstart = 1)

## create a new data frame for the ggplot visualisation 
gg_pca <- data.frame( PC1 = PC1,
                      PC2 = PC2, 
                      Cell_line = rownames(exp), 
                      Subtype = brain_cancer_cl$disease_subtype, 
                      Cluster = as.factor(km_c$cluster))

ggplot(gg_pca, aes( x = PC1, y = PC2, shape = Cluster, colour = Subtype)) +
  xlab("PC1") + 
  ylab("PC2")+ 
  geom_point(size = 4)

```
Cedrik beschreibt das plot

```{r eval=FALSE, echo=FALSE, fig.dim= c(5,4), fig.cap= 'Distribution of effectivness scores in each subtype'}
## create a new boxplot that shows the overall the distribution and 

p <- ggplot(new_means_eff, aes(x = subtype, y = Efficiency, color = subtype)) + 
                         ggtitle("Subtype Boxplots") +
                          xlab("Subtypes") + 
                          ylab("Efficiency") + 
  geom_boxplot(outlier.colour="red", outlier.shape=8,
                outlier.size=4)
p + theme(legend.position = "none")
```
IS THIS RELEVANT???

In these boxplots, one can see the distribution of the effectiveness scores of the 51 effective drugs in the subtypes. These can be used to compare the distributions of the effectiveness values for the subtypes. The median in astrocytoma, glioblastoma and glioma is in the middle of the box, which indicates a symmetrical distribution. 
The box of medulloblastoma differs from those of the other subtypes. It is larger and thus there is a greater spread of data points. In addition, the median is right-shifted and at a lower effective score. 
Although the boxplots provide information that the treatments worked differently effectively for the subtypes, they do not indicate whether there are drugs that worked well specifically for one subtype.  


+++


To find out which genes are expressed significantly different in brain cancer cell lines compared to other cancers, a statistical test can be applied. For this purpose, it must first be examined whether the available data is normally distributed.
A statistical test, the Shapiro Wilks test, can be used for this evaluation. In this case, the H0 hypothesis states that the distribution is a normal distribution. If the p-value is smaller than the significance level, the H0 hypothesis is rejected and it is assumed that the data is not normally distributed. 
The test was applied to the 19177 genes from the ```prism.exp``` dataset to determine the expression across the 34 cell lines. A signifcance level of 0.05 was set. As a result, 9799 p-values were less than or equal to this value, thus approximately half of the genes were normally distributed. This is the reason a non-parametric test was conducted.  

## Wilcoxon Rank Sum test

We applied an unpaired Wilcoxon Rank Sum test to identify differential expression. For this we created the dataset ```non_bc_exp```, which takes all cell lines from the ```prism.exp``` dataset that do not belong to the brain cancer disease type. There are 447 cell lines in it from lung cancer, skin cancer, pancreatic cancer or ovarian cancer. Although our data sets are not the same size, this is not a problem in this case, as the test is based on ranking the individual values.

Before applying the test, however, we had to take into account that we perform a Z-transformation by which the values from the two datasets are centered and scaled to match the different ranges of the variation. In addition, we had to take into account that for some genes the expression value was zero. Therefore, we added a so-called pseudocount, which was in our case the lowest expression value per data set. 


```{r echo=FALSE}
# Creating dataframe with all non brain cancer cell lines
non_bc_cl = prism.cl[!prism.cl$disease == "Brain Cancer",]
non_bc_names=non_bc_cl[,1] 
non_bc_exp=prism.exp[non_bc_names,]
brain_cancer_exp_w=brain_cancer_exp
non_bc_exp=non_bc_exp[rowSums(is.na(non_bc_exp)) == 0,]



#Z-transformation
calculate_pseudo_v<-function(x,y){
   min(x[y,][which(x[y,]>0)])
}
#function to determine the pseudovalue for both datasets. It takes a dataframe and looks for a minimum value for each row (in this case cell line). Then it chooses only the minimum values that are bigger than 0. Out of these positive mins, it chooses the smallest one, which is our pseudo value

pseudo_value1<-calculate_pseudo_v(non_bc_exp,nrow(non_bc_exp))
pseudo_value2<-calculate_pseudo_v(brain_cancer_exp_w, nrow(brain_cancer_exp_w))
#Now we have 2 pseudo values from 2 datasets, we wil compare them in the following loop to determine the ultimate pseudo value to apply on both datasets

if(pseudo_value1==pseudo_value2){
  pseudo_value<-pseudo_value1
} else if(pseudo_value1>pseudo_value2) {
  pseudo_value<-pseudo_value2
} else {
  pseudo_value<-pseudo_value1
}


non_bc_exp=non_bc_exp+pseudo_value
brain_cancer_exp_w=brain_cancer_exp_w+pseudo_value

non_bc_exp=as.data.frame(t(scale(as.matrix(t(non_bc_exp)))))
brain_cancer_exp_w=as.data.frame(t(scale(as.matrix(t(brain_cancer_exp_w)))))

# Removal of lines with NA from new dataframe
brain_cancer_exp_w=as.data.frame(t(na.omit(t(brain_cancer_exp_w))))
non_bc_exp=non_bc_exp[,colnames(brain_cancer_exp_w)]
```

```{r}

# Performance of Wilcoxon test

wilcox_genes <- vector(mode = "list",length=ncol(brain_cancer_exp_w))
for(i in 1:ncol(brain_cancer_exp_w)) {         
  wilcox_genes[[i]] <- wilcox.test(as.numeric(non_bc_exp[,i]),
                                   as.numeric(brain_cancer_exp_w[,i]), 
                                   alternative = "two.sided", paired = F)
}
```

When performing the Wilcoxon test between the two datasets ```non_bc_exp``` and ```brain_cancer_exp```, the H0 hypothesis that the gen expressions are from the same distribution is tested simultaneously for all genes. This leads to a multiplicity problem in which the probability of increasing the number of false positive test results is increased. To adjust the number of significant p-values for this problem, we consider two types of corrections. The Benjamini-Hochberg and the Bonferroni and corrections. 

Thereby, the number of genes that are significantly differential expressed differs strongly: The Benjamini-Hochberg correction results in 8948 genes, the Bonferroni correction in ***1859 genes***. We have chosen to continue with the Bonferroni correction genes, since this correction is more conservative and controls the type I error, so the number of false positives (keeping genes that are not significant) what we want prioritized to avoid. 

```{r echo=FALSE, fig.dim= c(5,4), fig.cap= 'Volcano plot of significant differential expressed genes'}

#P values of Wilcoxon test genes
wilcox_pvalue_genes <- vector(length = ncol(brain_cancer_exp_w))
for(i in 1:ncol(brain_cancer_exp_w)) {
  wilcox_pvalue_genes[i] <- wilcox_genes[[i]]$p.value
}

#Bonferroni correction
wilcox_pvalue_bonf = p.adjust(wilcox_pvalue_genes, method = "bonferroni", n = length(wilcox_pvalue_genes)) #I first was confused from p-value=1 but is possible when there is absolutely no evidence to reject H0
b_correction=which(wilcox_pvalue_bonf<=0.01)
p_values_b=wilcox_pvalue_bonf[b_correction]

non_bc_exp=prism.exp[non_bc_names,]

volcano_bc=brain_cancer_exp[,b_correction]
volcano_non_bc=non_bc_exp[,b_correction]

df<-data.frame(matrix(nrow = length(b_correction),ncol = 2))
rownames(df)=colnames(volcano_bc)
colnames(df)=c("-log10 p-value", "log2 foldchange")
df[,1]=-log10(p_values_b)

bc_exp_mean=apply(volcano_bc,2, mean)
non_bc_exp_mean=apply(volcano_non_bc,2,mean, na.rm=T)

df[,2]=log2((non_bc_exp_mean/bc_exp_mean))
logfc=df[,2]

highlight_df <- df %>% filter(logfc>=1|logfc<=-1)
bp<-sort(highlight_df[,1],decreasing = T)[1:20]
best_p_values<-data.frame(matrix(nrow =length(bp),ncol = 2))
best_p_values<-data.frame(matrix(nrow =length(bp),ncol = 2))
for (y in 1:length(bp)) {
  best_p_values[y,]<-highlight_df[which(highlight_df[,1]==bp[y]),]
  rownames(best_p_values)[y]=rownames(highlight_df[which(highlight_df[,1]==bp[y]),])
}


ggplot(df, aes(x=df[,2], y=df[,1]),ylim = range(0,1))+
  xlab("Log2-Fold-Change") + 
  ylab("-log10(p-value)") + 
  geom_point() +
  geom_point(data=highlight_df, aes(x=highlight_df[,2], y=highlight_df[,1]), color='blue') +
  geom_point(data=best_p_values, aes(x=best_p_values[,2], y=best_p_values[,1]), color='red')
```




To visualize differential expressed genes one can use a volcano plot. This plots the statistical significance of a gene in relation to the magnitude of the difference between the two datasets at a negative base 2 log-fold-change. Through this plot we could extract genes whose log2-fold-change is less than or equal to -1 and greater than or equal to 1, respectively. These genes are shown in blue in the plot. Additionally, we selected the top 20 genes with the smallest p-value, which are highlighted in red.

## Linear regression model with universal drug

In preparation for our final milestone, we wanted to further reduce the number of significant genes for brain cancer. 
For this, we wanted to extract a drug from our effective drugs in doses 7 that has a consistent effect for all brain cancer cell lines. Therefore, we filtered out an universal drug by dividing the range of effectiveness scores into intervals and then determining the interval with the highest number of cell lines responding to a treatment within all effective drugs. From this, it returned the drug with the Broad ID "BRD-K79145628-001-05-5". +++

We had a demand on this drug that it was the universal drug for brain cancer only, but not for other types of cancer. By applying our code to the non-brain cancer cell lines, our expectation was met. 

With this drug, we wanted to establish a linear regression model what is a statistical method to pedict one variable by using other variables. In our model, we predicted the effectiveness score of the universal drug based on the 20 genes filtered. By setting up a multiple regression, it is important to note that no correlated variables are used. Among our top 20 genes, we found a correlation between the genes "FAM83H" and "IQANK1". We kept the gene that has a more significant p-value in the Volcano plot. 

The application of the linear regression model involves feature selection, that is, reducing the number of genes so that only those that contribute to a significantly better prediction of effectiveness are included in the model. Here, the p-value of the F-test was relevant for our evaluation, since it compares our model with the null model. The F-test assumes the H0 hypotheses that the tested model does not perform better than the null model, so a small p-value leads to the rejection of the H0 hypotheses. 

We followed the strategy of removing as many genes from the model until the removal of one gene would make the p-value worse. At this event, the execution is stopped. Using this strategy, we arrive at a model with ***7 genes*** and a p-value of 1,353%. Moreover, in the output of the model, the multiple R-squared is relevant, as it indicates how much variance can be explained by the gene. This value is with 46,44% satisfactory for us.

To double check our result, we compared our complete model with the reduced model. For this we used the function ```anova```, which compares the variances of the residuals considering the introduced degrees of freedom. The H0 hypothesis is assumed that the two models are equivalent. Since a high p-value results, this can be rejected and it can be assumed that the complex model does not perform significantly better than our reduced model. 

+++Interpretation von Regressionsmodell auf alle cell lines?+++

# Linear regression model for effectivness prediction

Lea&Lennard

# Discussion



Literature:
- Global Cancer Statistics 2020
- Quelle zu subtypes
- Quelle zu Bonferroni vs BH
