---
title: "Search function"
author: "Lennard Kleemann"
date: "16/07/2021"
output: html_document
---

---
title: "Regression model"
author: "Lennard, Lea"
date: "13/7/2021"
output: html_document
---

```{r}
load("../2021-topic-01-team-03/dataset/prism_datasets.rda")
load("../2021-topic-01-team-03/dataset/cellline_datasets.rda")
```

# Guidelines

prism: effect of the treatment (columns) on cell growth of cell lines (rows).For example, BRD-A00077618-236-07-6::0.0024414::HTS002 indicates treatment with the drug BRD-A00077618-236-07-6, with a dose of 2.4 nM, in assay HTS002. Values represent log fold changes with respect to a control DMSO experiment.

prism.treat:information about the treatment (row) and the drug used

prism.cl: cell lines' info

prism.exp: gene TPM (transcripts per million) values. Indicate over/underexpression. Rows = cell lines, columns = genes

prism.cnv: gene copy number (CN) values. In absolute terms, CN = 2, since there are two alleles per genes. In cancer, genes might be amplified CN > 2 or deleted CN < 2

prism.snv: observed mutations in a sample. The isDeleterious flag specifies if mutation has a functional effect or not. First column is genes. There is a column
(DepMap_ID) with corresponding cell lines

prism.achilles: gene KO scores, a measure of how essential/important is a particular gene for the cell survival

IDEA: use same indications as in PRISM paper to start filtering for relevant genes, cell lines? 

# Packages 

```{r Packages}
# here is the place where we can put external packages

```

# Cleaning

## Selecting brain cancer specific cell lines in all data sets

```{r Selection of cell lines}
brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") #identify the brain cancer cell lines' names

names=brain_cancer_cl[,1] #saving the names as names

#select wanted cell lines from all datasets
brain_cancer=prism[names,]
brain_cancer_achilles=prism.achilles[names,]
brain_cancer_cnv=prism.cnv[names,]
brain_cancer_exp=prism.exp[names,]
```

## Working on NAs

```{r Working on NAs}
# replacing NAs 
lines=row.names(brain_cancer)
row.names(brain_cancer_achilles) = lines

# NA removal 
brain_cancer_achilles_no_NA=na.omit(brain_cancer_achilles) #remove the entire cell lines from everywhere without KO scores

# NA substitution for mean in brain_cancer
for(i in 1:ncol(brain_cancer)){
  brain_cancer[is.na(brain_cancer[,i]), i] <- mean(brain_cancer[,i], na.rm = TRUE)
}
brain_cancer=brain_cancer[ , colSums(is.na(brain_cancer)) == 0]
```

## Removal of useless columns 

```{r}
brain_cancer_cl=brain_cancer_cl[,-c(4,6,7,8,9,11,12,15,17,19,23,24,26,28,29,30,31,32,33)]
```



----------------------------------------
NEW CODE
----------------------------------------

## Improved regression model
```{r}

final_genes <- c("IQANK1","RAB11FIP1","FOXG1","LPIN3","ERVMER34.1","PTN","VRK2")

search.prism <- function(search.mode = "t", cellline, target.or.drug, dose = 8){

  #mode: targe=t, drug=d
  #target/drug:target name/ broad_id
  #dose: d1-d8
  doses.tabell <- data.frame()
  doses.tabell <- c(0,0.00061034,0.0024414,0.00976562,0.0390625,0.15625,0.625,2.5,10)
  bot.v <- vector()
  for(i in 1:c(length(doses.tabell)-1)){bot.v <- cbind(bot.v,c(doses.tabell[i]+(doses.tabell[c(i+1)]-doses.tabell[i])/2)) }
  bot.v <- cbind(bot.v, 10)
  doses.tabell <- rbind(doses.tabell,bot.v)
  doses.tabell[2,1] <- 0
  row.names(doses.tabell) <- c("Original dosis","End of intervall")
  colnames(doses.tabell) <- c("d0","d1","d2","d3","d4","d5","d6","d7","d8")

  
  ### Mode: target
  
  if(search.mode == "t" ){  
  
    
  target <- target.or.drug  
  
  res <- data.frame() #row.names = c(cellline, target))
  
  lts.line_gen <- prism.exp[grep(cellline,row.names(prism.exp)),which(colnames(prism.exp) %in% final_genes)] # selecting exp from 9 important gens
    
  lts.achilles <- prism.achilles[grep(cellline,row.names(prism.achilles)), which(target == colnames(prism.achilles))] # selecting achilles
  
  lts.exp <- prism.exp[grep(cellline,row.names(prism.exp)), which(target == colnames(prism.exp))] # selecting exp
  
  lts.cnv <- prism.cnv[grep(cellline,row.names(prism.cnv)), which(target == colnames(prism.cnv))] # selecting cnv
  
  ## target to drug to prism
  
  # finding drugs to target
  
  drug.line <- vector()
  
  for(i in 1:nrow(prism.treat)){
    
    target.in.line <- unlist(strsplit(as.character(prism.treat[i,7]), split = ", ")) == target # linewise target search in prism.treat
    
    if(sum(target.in.line, na.rm = TRUE) > 0){
      drug.line <- cbind(drug.line, i) # lines of drugs in prism.treat
    }
  }
  
  drugs.from.target <- unique(as.character(prism.treat[drug.line,1]))
  
  # finding prism values to drugs
  
  drug.col <- vector()
  
  for(i in 1:ncol(prism)){
    
    drug.in.col <- unlist(strsplit(as.character(colnames(prism)[i]), split = "::"))[1] == drugs.from.target # searching drugs in prism column
    
    if(sum(drug.in.col, na.rm = TRUE) > 0){
      drug.col <- cbind(drug.col, i) # columns with drugs that match target
    }
  }
  
  
  # dosis intervall
  lower.v <- doses.tabell[2,dose]
  upper.v <- doses.tabell[2,dose+1]
  
  # dosis check
  
  drug.with.dose <- vector()
  
  for(i in drug.col){
  
  drug.dose <- as.numeric(unlist(strsplit(as.character(colnames(prism)[i]), split = "::"))[2]) 
  
  if( lower.v < drug.dose && drug.dose <= upper.v ){
     
     drug.with.dose <- cbind(drug.with.dose,i) 
     
     } 
  }
  
  # prism values
  
  lts.prism <- mean(as.numeric(prism[grep(cellline,row.names(prism)), drug.with.dose]), na.rm = TRUE)
    
  # combine
  
  res <- cbind(lts.line_gen,lts.achilles,lts.exp, lts.cnv, lts.prism)
  
  row.names(res) <- paste(cellline, target)
  
  colnames(res) <- c(final_genes,"achilles","exp","cnv","prism")
  
  cat("Mode: cellline and target, cellline:",cellline, ", target:",target.or.drug,", dosis:", dose, "\n", "Drugs with target:", drugs.from.target, "\n")
  
  return(res)
  }
  
  
  
  
  ###Mode: drug
  
  else if(search.mode == "d"){
  

  drug <- target.or.drug
  
  res <- data.frame() #row.names = c(cellline, target))
  
  lts.line_gen <- prism.exp[grep(cellline,row.names(prism.exp)),which(colnames(prism.exp) %in% final_genes)] # selecting exp from 9 important gens
  
  #searching drug tagets
  
  targets <- as.character(prism.treat[grep(drug, prism.treat[,1]),7])
  
  multi.targets <- vector()
  
  for(i in targets){
    
    multi.targets <-c(multi.targets, unlist(strsplit(i,", ")))
  }
  
  targets <- unique(multi.targets)
  
  lts.achilles <- mean(as.numeric(prism.achilles[grep(cellline,row.names(prism.achilles)), which(colnames(prism.achilles) %in% targets)]), na.rm = TRUE) # selecting achilles
  
  lts.exp <- mean(as.numeric(prism.exp[grep(cellline,row.names(prism.exp)), which(colnames(prism.exp) %in% targets)]), na.rm = TRUE) # selecting exp
  
  lts.cnv <- mean(as.numeric(prism.cnv[grep(cellline,row.names(prism.cnv)), which(colnames(prism.cnv) %in% targets)]), na.rm = TRUE) # selecting cnv
  
 
  
  # dosis intervall
  lower.v <- doses.tabell[2,dose]
  upper.v <- doses.tabell[2,dose+1]
  
  # prism value
  
  drug.position <- vector()
  
  for(i in 1:ncol(prism)){
    
    act.drug <- unlist(strsplit(as.character(colnames(prism[i])), split = "::"))[1]
    act.dose <- as.numeric(unlist(strsplit(as.character(colnames(prism[i])), split = "::"))[2])
    
    if (act.drug == drug && lower.v < act.dose && act.dose <= upper.v){
      
      drug.position <- cbind(drug.position,i)
    }
    
  }
  
    lts.prism <- mean(as.numeric(prism[grep(cellline,row.names(prism)), drug.position]), na.rm = TRUE) 
    
  # combine
  
  res <- cbind(lts.line_gen,lts.achilles,lts.exp, lts.cnv, lts.prism)
  
  row.names(res) <- paste(cellline, "::", drug)
  
  colnames(res) <- c(final_genes,"achilles","exp","cnv","prism")
  
  cat("Mode: cellline and drug, cellline:",cellline, ", drug:",target.or.drug,", dosis:", dose, "\n", "Targets of the drug:", targets, "\n")
  
  return(res)
  }
  
  else{print("select s.mode")}
  
}


```







