---
title: "Search function"
author: "Lennard Kleemann"
date: "16/07/2021"
output: html_document
---

---
title: "Regression model"
author: "Lennard, Lea"
date: "13/7/2021"
output: html_document
---

```{r}
load("../2021-topic-01-team-03/dataset/prism_datasets.rda")
load("../2021-topic-01-team-03/dataset/cellline_datasets.rda")
```

# Guidelines

prism: effect of the treatment (columns) on cell growth of cell lines (rows).For example, BRD-A00077618-236-07-6::0.0024414::HTS002 indicates treatment with the drug BRD-A00077618-236-07-6, with a dose of 2.4 nM, in assay HTS002. Values represent log fold changes with respect to a control DMSO experiment.

prism.treat:information about the treatment (row) and the drug used

prism.cl: cell lines' info

prism.exp: gene TPM (transcripts per million) values. Indicate over/underexpression. Rows = cell lines, columns = genes

prism.cnv: gene copy number (CN) values. In absolute terms, CN = 2, since there are two alleles per genes. In cancer, genes might be amplified CN > 2 or deleted CN < 2

prism.snv: observed mutations in a sample. The isDeleterious flag specifies if mutation has a functional effect or not. First column is genes. There is a column
(DepMap_ID) with corresponding cell lines

prism.achilles: gene KO scores, a measure of how essential/important is a particular gene for the cell survival

IDEA: use same indications as in PRISM paper to start filtering for relevant genes, cell lines? 

# Packages 

```{r Packages}
# here is the place where we can put external packages

```


# Cleaning

## Selecting brain cancer specific cell lines in all data sets

```{r Selection of cell lines}
brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") #identify the brain cancer cell lines' names

names=brain_cancer_cl[,1] #saving the names as names

#select wanted cell lines from all datasets
brain_cancer=prism[names,]
brain_cancer_achilles=prism.achilles[names,]
brain_cancer_cnv=prism.cnv[names,]
brain_cancer_exp=prism.exp[names,]
```

## Working on NAs

```{r Working on NAs}
# replacing NAs 
lines=row.names(brain_cancer)
row.names(brain_cancer_achilles) = lines

# NA removal 
brain_cancer_achilles_no_NA=na.omit(brain_cancer_achilles) #remove the entire cell lines from everywhere without KO scores

# NA substitution for mean in brain_cancer
for(i in 1:ncol(brain_cancer)){
  brain_cancer[is.na(brain_cancer[,i]), i] <- mean(brain_cancer[,i], na.rm = TRUE)
}
brain_cancer=brain_cancer[ , colSums(is.na(brain_cancer)) == 0]
```

## Removal of useless columns 

```{r}
brain_cancer_cl=brain_cancer_cl[,-c(4,6,7,8,9,11,12,15,17,19,23,24,26,28,29,30,31,32,33)]
```

## Dividing cell lines into subtypes

```{r}
medulloblastoma=brain_cancer_cl$DepMap_ID[grep("Medulloblastoma", brain_cancer_cl$disease_subtype)]#we obtain the names of the cell lines that are classified as medulloblastoma

glioblastoma=brain_cancer_cl$DepMap_ID[grep("Glioblastoma", brain_cancer_cl$disease_subtype)]

glioma=brain_cancer_cl$DepMap_ID[grep("Glioma", brain_cancer_cl$disease_subtype)]

astrocytoma=brain_cancer_cl$DepMap_ID[grep("Astrocytoma", brain_cancer_cl$disease_subtype)]

#making sure we didnt forget anything 

disease_subtypes=c(medulloblastoma,glioblastoma,glioma,astrocytoma)
length(disease_subtypes)

# Converting them into dataframes [IS THIS RELEVANT?]
glioblastoma_bc <- brain_cancer[glioblastoma,]
medulloblastoma_bc <- brain_cancer[medulloblastoma,]
glioma_bc <- brain_cancer[glioma,]
astrocytoma_bc <- brain_cancer[astrocytoma,]
```


# Question 1: How can we distinguish the most effective drugs?

## Dividing drugs in doses groups

```{r}
dd1=grep("::0.00061034::", colnames(brain_cancer))
dd2=grep("::0.0024414::", colnames(brain_cancer))
dd3=grep("::0.00976562::", colnames(brain_cancer))
dd4=grep("::0.0390625::", colnames(brain_cancer))
dd5=grep("::0.15625::", colnames(brain_cancer))
dd6=grep("::0.625::", colnames(brain_cancer))
dd7=grep("::10::", colnames(brain_cancer))
dd8=grep("::2.5::", colnames(brain_cancer))
#positions of the respective doses in the dataframe 
doses=c(dd1,dd2,dd3,dd4,dd5,dd6,dd7,dd8)



doses_for_list=list("::0.00061034::","::0.0024414::","::0.00976562::","::0.0390625::","::0.15625::","::0.625::","::10::","::2.5::")
select_doses=function(x){
  brain_cancer[,grep(x, colnames(brain_cancer))]
}
doses_subsets_<-lapply(doses_for_list,select_doses)

check_for_differences<-function(x,y){
length(setdiff(colnames(x),colnames(y)))
}





d1=brain_cancer[,dd1]
d2=brain_cancer[,dd2]
d3=brain_cancer[,dd3]
d4=brain_cancer[,dd4]
d5=brain_cancer[,dd5]
d6=brain_cancer[,dd6]
d7=brain_cancer[,dd7]
d8=brain_cancer[,dd8]

check_for_differences(doses_subsets_[[1]],d1)
#these are the standard doses summed up into seperate dataframes
sum(length(dd1), length(dd2), length(dd3), length(dd4),length(dd5), length(dd6),length(dd7),length(dd8)) #ups something missing

deviation=brain_cancer[,-doses] #lets see what we missed-> these are the doses that are not exact and have some kind of a deviation

questionable_drugs=prism.treat[colnames(deviation),]#we dont know if we can use these drugs, may be their doses deviate too far from the standard
```

## Working on drugs that do not fit the standard doses

```{r}
#Stefan's suggestion: assigning the outliers to closest standard doses

#Calculation of all differences between questionable drugs and standard doses
std_doses <- c(0.00061034,0.0024414,0.00976562,0.0390625,0.15625,0.625,10,2.5) #vector with standard dosages

#Picking dosages from dosages that have to be assigned to standard dosages
doses_to_assign=as.numeric(questionable_drugs$dose)

#Creating data frame for the differences between standard dosages and questionable drugs
extra_drugs <-data.frame(matrix(nrow = 8,ncol = length(doses_to_assign)),row.names = std_doses)
colnames(extra_drugs)=rownames(questionable_drugs)

for (i in 1:length(doses_to_assign)){
    extra_drugs[,i]=abs(std_doses - doses_to_assign[i])
}

#Assigning the drugs to the closest standard dosage 
to_which_dose=apply(extra_drugs, 2, which.min) #function of which.min: gives row number of min value of column. In comparison just apply: gives min difference

d1_extra=which(to_which_dose==1)
d2_extra=which(to_which_dose==2)
d3_extra=which(to_which_dose==3)
d4_extra=which(to_which_dose==4)
d5_extra=which(to_which_dose==5)
d6_extra=which(to_which_dose==6)
d7_extra=which(to_which_dose==7)
d8_extra=which(to_which_dose==8)

doses_for_list_2<-list(1,2,3,4,5,6,7,8)
select_doses_2=function(x){
 unique(grep(x, to_which_dose))
}
doses_subsets_2<-lapply(doses_for_list_2,select_doses_2)

check_for_differences(doses_subsets_2[[7]],d7_extra)

add_extra_doses_list=function(x){
  cbind(brain_cancer[,unique(grep(paste(colnames(extra_drugs[,doses_subsets_2[[x]]]),collapse="|"), colnames(brain_cancer)))], doses_subsets_[[x]])
}
doses_subsets_<-lapply(doses_for_list_2,add_extra_doses_list)

#just a quick check how many doses there are in total
not_standard_doses=c(d1_extra,d2_extra,d3_extra,d4_extra,d5_extra,d6_extra,d7_extra,d8_extra)
length(not_standard_doses)#in comparison with dimension from questionable_drugs right!

#Adding new assigned dosages to small data frames of standard dosages 
add_extra_doses=function(x,y){
  cbind(brain_cancer[,unique(grep(paste(colnames(extra_drugs[,x]),collapse="|"), colnames(brain_cancer)))], y)
  }

d1=add_extra_doses(d1_extra,d1)
d2=add_extra_doses(d2_extra,d2)
d3=add_extra_doses(d3_extra,d3)
d4=add_extra_doses(d4_extra,d4)
d5=add_extra_doses(d5_extra,d5)
d6=add_extra_doses(d6_extra,d6)
d7=add_extra_doses(d7_extra,d7)
d8=add_extra_doses(d8_extra,d8)

#ResumÃ©: new data sets for each dosage with assigned outliers to closest deviation

ncol(d2)+ncol(d1)+ncol(d3)+ncol(d4)+ncol(d5)+ncol(d6)+ncol(d7)+ncol(d8)
ncol(brain_cancer)


check_for_differences(doses_subsets_[[8]],d8)
```

## Filter effective drugs in each dose group

```{r}
# Defining a threshold by looking at the data set
median(apply(brain_cancer,2,mean))


# Lets select the drugs that go with the threshold of 0.2 in each dose group
effective_drugs<-function(x){
  effective_x<-x[,which(apply(x,2,mean)<=0.2)]
  return(effective_x)
}

effective_d1=effective_drugs(d1)
effective_d2=effective_drugs(d2)
effective_d3=effective_drugs(d3)
effective_d4=effective_drugs(d4)
effective_d5=effective_drugs(d5)
effective_d6=effective_drugs(d6)
effective_d7=effective_drugs(d7)
effective_d8=effective_drugs(d8)



doses_subsets <- list(d1 = d1, d2 = d2, d3 = d3, d4 = d4, d5 = d5, d6 = d6, d7 = d7, d8 = d8)
all_effective_drugs_2<- lapply(doses_subsets, effective_drugs)
all_effective_drugs_<-lapply(doses_subsets_, effective_drugs)
check_for_differences(all_effective_drugs_2[[3]], all_effective_drugs_[[3]])

all_effective_drugs<-lapply(doses_subsets_, effective_drugs)


ncol(all_effective_drugs[[1]])+ncol(all_effective_drugs[[2]])+ncol(all_effective_drugs[[3]])+ncol(all_effective_drugs[[4]])+ncol(all_effective_drugs[[5]])+ncol(all_effective_drugs[[6]])+ncol(all_effective_drugs[[7]])+ncol(all_effective_drugs[[8]])

length(all_effective_drugs)
length(which(apply(brain_cancer,2,mean)<=0.2)) #this is the amount of drugs that we keep in total. We compare these 2 length to make sure that that we do not select the same drug twice. If it was the case the first length would be bigger than the second one
```

##Select drugs that are effective in all doses

```{r}
save_new_colnames<-function(x){
  a<-gsub(pattern = "::.*", replacement = "", colnames(x))
  colnames(x)<-a
  return(x)
}

effective_d1<-save_new_colnames(effective_d1)
effective_d2<-save_new_colnames(effective_d2)
effective_d3<-save_new_colnames(effective_d3)
effective_d4<-save_new_colnames(effective_d4)
effective_d5<-save_new_colnames(effective_d5)
effective_d6<-save_new_colnames(effective_d6)
effective_d7<-save_new_colnames(effective_d7)
effective_d8<-save_new_colnames(effective_d8)



all_effective_drugs<-lapply(all_effective_drugs, save_new_colnames)
check_for_differences(all_effective_drugs[[4]],effective_d4)
## remove everything but the drug information from colnames and make them new colnames of the dataframes

common_drugs<-function(y){
  different<-setdiff(colnames(effective_d1), colnames(y))
  
  different<-unique(grep(paste(different,collapse="|"), colnames(effective_d1)))
  
  colnames(effective_d1[,-different])
}

common_drugs_<-function(y){
  different<-setdiff(colnames(all_effective_drugs[[1]]), colnames(all_effective_drugs[[y]]))
  
  different<-unique(grep(paste(different,collapse="|"), colnames(all_effective_drugs[[1]])))
  
  colnames(all_effective_drugs[[1]][,-different])
}


#function returns the names of the common drugs between 2 dataframes

d1_d2=common_drugs(effective_d2)
#both are equal, so the function works correctly

d1_d3=common_drugs(effective_d3)
d1_d4=common_drugs(effective_d4)
d1_d5=common_drugs(effective_d5)
d1_d6=common_drugs(effective_d6)
d1_d7=common_drugs(effective_d7)
d1_d8=common_drugs(effective_d8)

doses_interception<-lapply(doses_for_list_2, common_drugs_)
length(setdiff(doses_interception[[5]],d1_d5))

#which drugs are present in all 8 doses?
present_in_all_doses_<-function(a,b,c,d,e,f,g){
 bb<-setdiff(a, b)
 a<-a [! a %in% bb]
 
 cc<-setdiff(a, c)
 a<-a [! a %in% cc]
 
 dd<-setdiff(a, d)
 a<-a [! a %in% dd]
 
 ee<-setdiff(a, e)
 a<-a [! a %in% ee]
 
 ff<-setdiff(a, f)
 a<-a [! a %in% ff]
 
 gg<-setdiff(a, g)
 a<-a [! a %in% gg]
 
 return(a)
}

present_in_all_doses<-function(a,b,c,d,e,f,g){
 bb<-setdiff(a, b)
 a<-a [! a %in% bb]
 
 cc<-setdiff(a, c)
 a<-a [! a %in% cc]
 
 dd<-setdiff(a, d)
 a<-a [! a %in% dd]
 
 ee<-setdiff(a, e)
 a<-a [! a %in% ee]
 
 ff<-setdiff(a, f)
 a<-a [! a %in% ff]
 
 gg<-setdiff(a, g)
 a<-a [! a %in% gg]
 
 return(a)
}

effective_in_all_doses=present_in_all_doses(d1_d2,d1_d3,d1_d4,d1_d5,d1_d6,d1_d7,d1_d8)
#51 drugs are effective in all doses

```

----------------------------------------
SEARCH FUNCTION (Because of two nested loops it picks out more than 33000 values, which will take some time!)
----------------------------------------

```{r}

final_genes <- c("IQANK1","RAB11FIP1","FOXG1","LPIN3","ERVMER34.1","PTN","VRK2")

search.prism <- function(search.mode = "t", cellline, target.or.drug, dose = 8){

  #mode: targe=t, drug=d
  #target/drug:target name/ broad_id
  #dose: d1-d8
  doses.tabell <- data.frame()
  doses.tabell <- c(0,0.00061034,0.0024414,0.00976562,0.0390625,0.15625,0.625,2.5,10)
  bot.v <- vector()
  for(i in 1:c(length(doses.tabell)-1)){bot.v <- cbind(bot.v,c(doses.tabell[i]+(doses.tabell[c(i+1)]-doses.tabell[i])/2)) }
  bot.v <- cbind(bot.v, 10)
  doses.tabell <- rbind(doses.tabell,bot.v)
  doses.tabell[2,1] <- 0
  row.names(doses.tabell) <- c("Original dosis","End of intervall")
  colnames(doses.tabell) <- c("d0","d1","d2","d3","d4","d5","d6","d7","d8")

  
  ### Mode: target
  
  if(search.mode == "t" ){  
  
    
  target <- target.or.drug  
  
  res <- data.frame() #row.names = c(cellline, target))
  
  lts.line_gen <- prism.exp[grep(cellline,row.names(prism.exp)),which(colnames(prism.exp) %in% final_genes)] # selecting exp from 9 important gens
    
  lts.achilles <- prism.achilles[grep(cellline,row.names(prism.achilles)), which(target == colnames(prism.achilles))] # selecting achilles
  
  lts.exp <- prism.exp[grep(cellline,row.names(prism.exp)), which(target == colnames(prism.exp))] # selecting exp
  
  lts.cnv <- prism.cnv[grep(cellline,row.names(prism.cnv)), which(target == colnames(prism.cnv))] # selecting cnv
  
  ## target to drug to prism
  
  # finding drugs to target
  
  drug.line <- vector()
  
  for(i in 1:nrow(prism.treat)){
    
    target.in.line <- unlist(strsplit(as.character(prism.treat[i,7]), split = ", ")) == target # linewise target search in prism.treat
    
    if(sum(target.in.line, na.rm = TRUE) > 0){
      drug.line <- cbind(drug.line, i) # lines of drugs in prism.treat
    }
  }
  
  drugs.from.target <- unique(as.character(prism.treat[drug.line,1]))
  
  # finding prism values to drugs
  
  drug.col <- vector()
  
  for(i in 1:ncol(prism)){
    
    drug.in.col <- unlist(strsplit(as.character(colnames(prism)[i]), split = "::"))[1] == drugs.from.target # searching drugs in prism column
    
    if(sum(drug.in.col, na.rm = TRUE) > 0){
      drug.col <- cbind(drug.col, i) # columns with drugs that match target
    }
  }
  
  
  # dosis intervall
  lower.v <- doses.tabell[2,dose]
  upper.v <- doses.tabell[2,dose+1]
  
  # dosis check
  
  drug.with.dose <- vector()
  
  for(i in drug.col){
  
  drug.dose <- as.numeric(unlist(strsplit(as.character(colnames(prism)[i]), split = "::"))[2]) 
  
  if( lower.v < drug.dose && drug.dose <= upper.v ){
     
     drug.with.dose <- cbind(drug.with.dose,i) 
     
     } 
  }
  
  # prism values
  
  lts.prism <- mean(as.numeric(prism[grep(cellline,row.names(prism)), drug.with.dose]), na.rm = TRUE)
    
  # combine
  
  res <- cbind(lts.line_gen,lts.achilles,lts.exp, lts.cnv, lts.prism)
  
  row.names(res) <- paste(cellline, target)
  
  colnames(res) <- c(final_genes,"achilles","exp","cnv","prism")
  
  cat("Mode: cellline and target, cellline:",cellline, ", target:",target.or.drug,", dosis:", dose, "\n", "Drugs with target:", drugs.from.target, "\n")
  
  return(res)
  }
  
  
  
  
  ###Mode: drug
  
  else if(search.mode == "d"){
  

  drug <- target.or.drug
  
  res <- data.frame() #row.names = c(cellline, target))
  
  lts.line_gen <- prism.exp[grep(cellline,row.names(prism.exp)),which(colnames(prism.exp) %in% final_genes)] # selecting exp from 9 important gens
  
  #searching drug tagets
  
  targets <- as.character(prism.treat[grep(drug, prism.treat[,1]),7])
  
  multi.targets <- vector()
  
  for(i in targets){
    
    multi.targets <-c(multi.targets, unlist(strsplit(i,", ")))
  }
  
  targets <- unique(multi.targets)
  
  lts.achilles <- mean(as.numeric(prism.achilles[grep(cellline,row.names(prism.achilles)), which(colnames(prism.achilles) %in% targets)]), na.rm = TRUE) # selecting achilles
  
  lts.exp <- mean(as.numeric(prism.exp[grep(cellline,row.names(prism.exp)), which(colnames(prism.exp) %in% targets)]), na.rm = TRUE) # selecting exp
  
  lts.cnv <- mean(as.numeric(prism.cnv[grep(cellline,row.names(prism.cnv)), which(colnames(prism.cnv) %in% targets)]), na.rm = TRUE) # selecting cnv
  
 
  
  # dosis intervall
  lower.v <- doses.tabell[2,dose]
  upper.v <- doses.tabell[2,dose+1]
  
  # prism value
  
  drug.position <- vector()
  
  for(i in 1:ncol(prism)){
    
    act.drug <- unlist(strsplit(as.character(colnames(prism[i])), split = "::"))[1]
    act.dose <- as.numeric(unlist(strsplit(as.character(colnames(prism[i])), split = "::"))[2])
    
    if (act.drug == drug && lower.v < act.dose && act.dose <= upper.v){
      
      drug.position <- cbind(drug.position,i)
    }
    
  }
  
    lts.prism <- mean(as.numeric(prism[grep(cellline,row.names(prism)), drug.position]), na.rm = TRUE) 
    
  # combine
  
  res <- cbind(lts.line_gen,lts.achilles,lts.exp, lts.cnv, lts.prism)
  
  row.names(res) <- paste(cellline, "::", drug)
  
  colnames(res) <- c(final_genes,"achilles","exp","cnv","prism")
  
  cat("Mode: cellline and drug, cellline:",cellline, ", drug:",target.or.drug,", dosis:", dose, "\n", "Targets of the drug:", targets, "\n")
  
  return(res)
  }
  
  else{print("select s.mode")}
  
}


```

# Selecting data with cell lines and targets
```{r}

training_targets <- brain_cancer_treat[effective_in_all_doses,7]
training_targets <- as.character(training_targets)
x <- 1

training_targets.solo <- list() #create list

for(i in training_targets) #split multiple targets
  {
  
 training_targets.solo[[x]] <- strsplit(i, split = ", ")
 x <- x+1
 }
training_targets.solo <- unlist(training_targets.solo) #list to vector

training_targets.solo <- unique(training_targets.solo) # remove multiple targets

training_targets.solo <- training_targets.solo[-1] # remove NA

training_targets.solo <- sort(training_targets.solo) # sort alphabetic

training_targets.solo <- training_targets.solo[-which(training_targets.solo %in% c("ATP5A1","MMP12","MMP23A"))]#remove targets with incomplete data


celllines <- brain_cancer_achilles[,1]
celllines <- cbind(celllines, rownames(brain_cancer_achilles))
celllines <- na.omit(celllines)

training <- data.frame()

for (i in celllines[,2]){
  print(i)
  for (j in training_targets.solo){
    
    training <- rbind(training, search.prism("t",i,j,8))
  }
}


training <- na.omit(training)# final training dataset for linear model

# save(training, file = "linear.model_data.RData") # saving if wanted 
```





